// cspell-io/dist/cjs/node/file/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStatSync = exports.getStat = exports.writeToFileIterable = exports.writeToFile = exports.readFileSync = exports.readFile = void 0;
var fileReader_js_1 = require("./fileReader.js");
Object.defineProperty(exports, "readFile", { enumerable: true, get: function () { return fileReader_js_1.readFile; } });
Object.defineProperty(exports, "readFileSync", { enumerable: true, get: function () { return fileReader_js_1.readFileSync; } });
var fileWriter_js_1 = require("./fileWriter.js");
Object.defineProperty(exports, "writeToFile", { enumerable: true, get: function () { return fileWriter_js_1.writeToFile; } });
Object.defineProperty(exports, "writeToFileIterable", { enumerable: true, get: function () { return fileWriter_js_1.writeToFileIterable; } });
var stat_js_1 = require("./stat.js");
Object.defineProperty(exports, "getStat", { enumerable: true, get: function () { return stat_js_1.getStat; } });
Object.defineProperty(exports, "getStatSync", { enumerable: true, get: function () { return stat_js_1.getStatSync; } });
