// cspell-lib/dist/textValidation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateText = exports.calcTextInclusionRanges = exports.DocumentValidator = exports.IncludeExcludeFlag = exports.checkTextDocument = exports.checkText = void 0;
var checkText_1 = require("./checkText");
Object.defineProperty(exports, "checkText", { enumerable: true, get: function () { return checkText_1.checkText; } });
Object.defineProperty(exports, "checkTextDocument", { enumerable: true, get: function () { return checkText_1.checkTextDocument; } });
Object.defineProperty(exports, "IncludeExcludeFlag", { enumerable: true, get: function () { return checkText_1.IncludeExcludeFlag; } });
var docValidator_1 = require("./docValidator");
Object.defineProperty(exports, "DocumentValidator", { enumerable: true, get: function () { return docValidator_1.DocumentValidator; } });
var textValidator_1 = require("./textValidator");
Object.defineProperty(exports, "calcTextInclusionRanges", { enumerable: true, get: function () { return textValidator_1.calcTextInclusionRanges; } });
var validator_1 = require("./validator");
Object.defineProperty(exports, "validateText", { enumerable: true, get: function () { return validator_1.validateText; } });
