// cspell-lib/dist/textValidation/parsedText.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapRangeBackToOriginalPos = exports.mapIssueBackToOriginalPos = void 0;
function mapIssueBackToOriginalPos(parsedText, issue) {
    if (!parsedText.map || parsedText.map.length === 0)
        return issue;
    const textOff = mapTextOffsetBackToOriginalPos(parsedText, issue);
    return {
        ...issue,
        ...textOff,
    };
}
exports.mapIssueBackToOriginalPos = mapIssueBackToOriginalPos;
function mapTextOffsetBackToOriginalPos(parsedText, textOff) {
    var _a;
    if (!parsedText.map || !parsedText.map.length)
        return textOff;
    const off = textOff.offset - parsedText.range[0];
    const range = mapRangeBackToOriginalPos([off, off + ((_a = textOff.length) !== null && _a !== void 0 ? _a : textOff.text.length)], parsedText.map);
    return {
        text: textOff.text,
        offset: parsedText.range[0] + range[0],
        length: range[1] - range[0],
    };
}
function mapRangeBackToOriginalPos(offRange, map) {
    if (!map || !map.length)
        return offRange;
    const [start, end] = offRange;
    let i = 0, j = 0, p = 1;
    while (p < map.length && map[p] < start) {
        i = map[p - 1];
        j = map[p];
        p += 2;
    }
    const iA = start - j + i;
    while (p < map.length && map[p] < end) {
        i = map[p - 1];
        j = map[p];
        p += 2;
    }
    const iB = end - j + i;
    return [iA, iB];
}
exports.mapRangeBackToOriginalPos = mapRangeBackToOriginalPos;
