// gensequence/dist/builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = void 0;
const ImplSequenceBuilder_js_1 = require("./ImplSequenceBuilder.js");
const index_js_1 = require("./operators/index.js");
function makeBuilder(fn) {
    return new ImplSequenceBuilder_js_1.ImplSequenceBuilder([fn]);
}
exports.builder = Object.freeze({
    pipe: (fn) => {
        return makeBuilder(fn);
    },
    //// Filters
    /** keep values where the fnFilter(t) returns true */
    filter: (fnFilter) => {
        return makeBuilder((0, index_js_1.filter)(fnFilter));
    },
    skip: (n) => {
        return makeBuilder((0, index_js_1.skip)(n));
    },
    take: (n) => {
        return makeBuilder((0, index_js_1.take)(n));
    },
    //// Extenders
    concat: (j) => {
        return makeBuilder((0, index_js_1.concat)(j));
    },
    concatMap: (fn) => {
        return makeBuilder((0, index_js_1.concatMap)(fn));
    },
    //// Mappers
    combine: (fn, j) => {
        return makeBuilder((0, index_js_1.combine)(fn, j));
    },
    /** map values from type T to type U */
    map: (fnMap) => {
        return makeBuilder((0, index_js_1.map)(fnMap));
    },
    scan: (fnReduce, initialValue) => {
        return makeBuilder((0, index_js_1.scan)(fnReduce, initialValue));
    },
});
