// cspell/dist/cjs/emitters/traceEmitter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.__testing__ = exports.emitTraceResults = void 0;
const chalk_1 = __importDefault(require("chalk"));
const iPath = __importStar(require("path"));
const strip_ansi_1 = __importDefault(require("strip-ansi"));
const util_js_1 = require("../util/util.js");
const colWidthDictionaryName = 20;
function emitTraceResults(results, options) {
    const maxWordLength = results
        .map((r) => r.foundWord || r.word)
        .reduce((a, b) => Math.max(a, (0, util_js_1.width)(b)), 'Word'.length);
    const maxDictNameLength = results
        .map((r) => r.dictName.length)
        .reduce((a, b) => Math.max(a, b), colWidthDictionaryName);
    const cols = {
        word: maxWordLength,
        dictName: maxDictNameLength,
        terminalWidth: options.lineWidth ?? (process.stdout.columns || 120),
        location: options.dictionaryPathFormat === 'hide' ? 0 : 30,
    };
    const col = new Intl.Collator();
    results.sort((a, b) => col.compare(a.dictName, b.dictName));
    emitHeader(cols);
    results.forEach((r) => emitTraceResult(r, cols, options));
}
exports.emitTraceResults = emitTraceResults;
function emitHeader(colWidths) {
    const line = [
        (0, util_js_1.pad)('Word', colWidths.word),
        'F',
        (0, util_js_1.pad)('Dictionary', colWidths.dictName),
        colWidths.location ? (0, util_js_1.pad)('Dictionary Location', colWidths.location) : '',
    ];
    console.log(chalk_1.default.underline(line.join(' ').trim().slice(0, colWidths.terminalWidth)));
}
function emitTraceResult(r, colWidths, options) {
    const { word: wordColWidth, terminalWidth, dictName: widthName } = colWidths;
    const errors = r.errors?.map((e) => e.message)?.join('\n\t') || '';
    const word = (0, util_js_1.pad)(r.foundWord || r.word, wordColWidth);
    const cWord = word.replace(/[+]/g, chalk_1.default.yellow('+'));
    const w = r.forbidden ? chalk_1.default.red(cWord) : chalk_1.default.green(cWord);
    const f = calcFoundChar(r);
    const a = r.dictActive ? '*' : ' ';
    const dictName = (0, util_js_1.pad)(r.dictName.slice(0, widthName - 1) + a, widthName);
    const dictColor = r.dictActive ? chalk_1.default.yellowBright : chalk_1.default.rgb(200, 128, 50);
    const n = dictColor(dictName);
    const info = [w, f, n].join(' ') + ' ';
    const used = (0, util_js_1.width)((0, strip_ansi_1.default)(info));
    const widthSrc = terminalWidth - used;
    const c = colorize(errors ? chalk_1.default.red : chalk_1.default.white);
    const s = c(formatDictionaryLocation(r.dictSource, widthSrc, { iPath, ...options }));
    const line = info + s;
    console.log(line.trim());
    if (errors) {
        console.error('\t' + chalk_1.default.red(errors));
    }
}
function trimMid(s, w) {
    s = s.trim();
    if (s.length <= w) {
        return s;
    }
    const l = Math.floor((w - 3) / 2);
    const r = Math.ceil((w - 3) / 2);
    return s.slice(0, l) + '...' + s.slice(-r);
}
function calcFoundChar(r) {
    const errors = r.errors?.map((e) => e.message)?.join('\n\t') || '';
    let color = chalk_1.default.dim;
    color = r.found ? chalk_1.default.whiteBright : color;
    color = r.forbidden ? chalk_1.default.red : color;
    color = r.noSuggest ? chalk_1.default.yellowBright : color;
    color = errors ? chalk_1.default.red : color;
    let char = '-';
    char = r.found ? '*' : char;
    char = r.forbidden ? '!' : char;
    char = r.noSuggest ? 'I' : char;
    char = errors ? 'X' : char;
    return color(char);
}
function formatDictionaryLocation(dictSource, maxWidth, { cwd, dictionaryPathFormat: format, iPath, }) {
    let relPath = cwd ? iPath.relative(cwd, dictSource) : dictSource;
    const idxNodeModule = relPath.lastIndexOf('node_modules');
    const isNodeModule = idxNodeModule >= 0;
    if (format === 'hide')
        return '';
    if (format === 'short') {
        const prefix = isNodeModule
            ? '[node_modules]/'
            : relPath.startsWith('..' + iPath.sep + '..')
                ? '.../'
                : relPath.startsWith('..' + iPath.sep)
                    ? '../'
                    : '';
        return prefix + iPath.basename(dictSource);
    }
    if (format === 'full')
        return dictSource;
    relPath = isNodeModule ? relPath.slice(idxNodeModule) : relPath;
    const usePath = relPath.length < dictSource.length ? relPath : dictSource;
    return trimMidPath(usePath, maxWidth, iPath.sep);
}
function colorize(fn) {
    return (s) => (s ? fn(s) : '');
}
function trimMidPath(s, w, sep) {
    if (s.length <= w)
        return s;
    const parts = s.split(sep);
    if (parts[parts.length - 1].length > w)
        return trimMid(s, w);
    function join(left, right) {
        // if (left === right) return parts.join(sep);
        return [...parts.slice(0, left), '...', ...parts.slice(right)].join(sep);
    }
    let left = 0, right = parts.length, last = '';
    for (let i = 0; i < parts.length; ++i) {
        const incLeft = i & 1 ? 1 : 0;
        const incRight = incLeft ? 0 : -1;
        const next = join(left + incLeft, right + incRight);
        if (next.length > w)
            break;
        left += incLeft;
        right += incRight;
        last = next;
    }
    for (let i = left + 1; i < right; ++i) {
        const next = join(i, right);
        if (next.length > w)
            break;
        last = next;
    }
    for (let i = right - 1; i > left; --i) {
        const next = join(left, i);
        if (next.length > w)
            break;
        last = next;
    }
    return last || trimMid(s, w);
}
exports.__testing__ = {
    trimMidPath,
};
