// gensequence/dist/operators/operatorsBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scanMap = exports.isAsyncIterableIterator = exports.isAsyncIterable = exports.makeAsyncIterable = exports.isIterableIterator = exports.isIterable = exports.makeIterable = exports.reduceAsyncForAsyncIterator = exports.reduceAsync = exports.reduce = exports.min = exports.max = exports.forEach = exports.first = exports.count = exports.any = exports.all = exports.scan = exports.map = exports.combine = exports.concatMap = exports.concat = exports.take = exports.skip = exports.filter = void 0;
/**
 * Operators used by Sequence
 */
//// Filters
function* filter(i, fnFilter) {
    for (const v of i) {
        if (fnFilter(v)) {
            yield v;
        }
    }
}
exports.filter = filter;
function* skip(i, n) {
    let a = 0;
    for (const t of i) {
        if (a >= n) {
            yield t;
        }
        a += 1;
    }
}
exports.skip = skip;
function* take(i, n) {
    let a = 0;
    if (n) {
        for (const t of i) {
            if (a >= n) {
                break;
            }
            yield t;
            a += 1;
        }
    }
}
exports.take = take;
//// Extenders
/**
 * Concat two iterables together
 */
function* concat(i, j) {
    yield* i;
    yield* j;
}
exports.concat = concat;
function* concatMap(i, fn) {
    for (const t of i) {
        yield* fn(t);
    }
}
exports.concatMap = concatMap;
//// Mappers
/**
 * Combine two iterables together using fnMap function.
 */
function* combine(i, j, fnMap) {
    const jit = j[Symbol.iterator]();
    for (const r of i) {
        const s = jit.next().value;
        yield fnMap(r, s);
    }
}
exports.combine = combine;
/**
 * apply a mapping function to an Iterable.
 */
function map(i, fnMap) {
    function* fn(i, fnMap) {
        for (const v of i) {
            yield fnMap(v);
        }
    }
    return fn(i, fnMap);
}
exports.map = map;
function* scan(i, fnReduce, initValue) {
    let index = 0;
    if (initValue === undefined) {
        // We need to create a new iterable to prevent for...of from restarting an array.
        index = 1;
        const iter = i[Symbol.iterator]();
        let r = iter.next();
        if (!r.done)
            yield r.value;
        initValue = r.value;
        i = makeIterable(iter);
    }
    let prevValue = initValue;
    for (const t of i) {
        const nextValue = fnReduce(prevValue, t, index);
        yield nextValue;
        prevValue = nextValue;
        index += 1;
    }
}
exports.scan = scan;
//// Reducers
function all(i, fn) {
    for (const t of i) {
        if (!fn(t)) {
            return false;
        }
    }
    return true;
}
exports.all = all;
function any(i, fn) {
    for (const t of i) {
        if (fn(t)) {
            return true;
        }
    }
    return false;
}
exports.any = any;
function count(i) {
    return reduce(i, (p) => p + 1, 0);
}
exports.count = count;
function first(i, fn, defaultValue) {
    fn = fn || (() => true);
    for (const t of i) {
        if (fn(t)) {
            return t;
        }
    }
    return defaultValue;
}
exports.first = first;
function forEach(i, fn) {
    let index = 0;
    for (const t of i) {
        fn(t, index);
        index += 1;
    }
}
exports.forEach = forEach;
function max(i, selector = (t) => t) {
    return reduce(i, (p, c) => (selector(c) > selector(p) ? c : p), undefined);
}
exports.max = max;
function min(i, selector = (t) => t) {
    return reduce(i, (p, c) => (selector(c) < selector(p) ? c : p), undefined);
}
exports.min = min;
function reduce(i, fnReduce, initialValue) {
    // We need to create a new iterable to prevent for...of from restarting an array.
    const iter = makeIterable(i[Symbol.iterator]());
    let index = 0;
    if (initialValue === undefined) {
        index = 1;
        const r = iter.next();
        initialValue = r.value;
    }
    let prevValue = initialValue;
    for (const t of iter) {
        const nextValue = fnReduce(prevValue, t, index);
        prevValue = nextValue;
        index += 1;
    }
    return prevValue;
}
exports.reduce = reduce;
async function reduceAsync(i, fnReduce, initialValue) {
    // We need to create a new iterable to prevent for...of from restarting an array.
    const iter = makeIterable(i[Symbol.iterator]());
    let index = 0;
    if (initialValue === undefined) {
        index = 1;
        const r = iter.next();
        initialValue = r.value;
    }
    let previousValue = await initialValue;
    for (const p of iter) {
        const t = await p;
        const nextValue = await fnReduce(previousValue, t, index);
        previousValue = nextValue;
        index += 1;
    }
    return previousValue;
}
exports.reduceAsync = reduceAsync;
async function reduceAsyncForAsyncIterator(i, fnReduce, initialValue) {
    const iter = makeAsyncIterable(i[Symbol.asyncIterator]());
    let index = 0;
    if (initialValue === undefined) {
        index = 1;
        const r = await iter.next();
        initialValue = r.value;
    }
    let previousValue = await initialValue;
    for await (const t of iter) {
        const nextValue = await fnReduce(previousValue, t, index);
        previousValue = nextValue;
        index += 1;
    }
    return previousValue;
}
exports.reduceAsyncForAsyncIterator = reduceAsyncForAsyncIterator;
//// Utilities
/**
 * Convert an Iterator into an IterableIterator
 */
function makeIterable(i) {
    function* fromIterator(i) {
        for (let r = i.next(); !r.done; r = i.next()) {
            yield r.value;
        }
    }
    function* fromIterable(i) {
        yield* i;
    }
    return isIterable(i) ? (isIterableIterator(i) ? i : fromIterable(i)) : fromIterator(i);
}
exports.makeIterable = makeIterable;
function isIterable(i) {
    return !!i[Symbol.iterator];
}
exports.isIterable = isIterable;
function isIterableIterator(i) {
    return typeof i.next == 'function';
}
exports.isIterableIterator = isIterableIterator;
function makeAsyncIterable(i) {
    async function* fromIterable(i) {
        for (const v of i) {
            yield v;
        }
    }
    async function* fromIterator(i) {
        for (let r = await i.next(); !r.done; r = await i.next()) {
            yield r.value;
        }
    }
    async function* fromAsyncIterable(i) {
        yield* i;
    }
    return isAsyncIterable(i) ? (isAsyncIterableIterator(i) ? i : fromAsyncIterable(i)) : isIterable(i) ? fromIterable(i) : fromIterator(i);
}
exports.makeAsyncIterable = makeAsyncIterable;
function isAsyncIterable(i) {
    return !!i[Symbol.asyncIterator];
}
exports.isAsyncIterable = isAsyncIterable;
function isAsyncIterableIterator(i) {
    return typeof i.next == 'function';
}
exports.isAsyncIterableIterator = isAsyncIterableIterator;
function scanMap(accFn, init) {
    let acc = init;
    let first = true;
    return function (value) {
        if (first && acc === undefined) {
            first = false;
            acc = value;
            return acc;
        }
        acc = accFn(acc, value);
        return acc;
    };
}
exports.scanMap = scanMap;
