// cspell-dictionary/dist/SpellingDictionary/SuggestDictionary.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSuggestDictionary = void 0;
const sync_1 = require("@cspell/cspell-pipe/sync");
const AutoResolve_1 = require("../util/AutoResolve");
const textMappers_1 = require("../util/textMappers");
const defaults = __importStar(require("./defaults"));
const Typos_1 = require("./Typos");
const util_1 = require("./Typos/util");
class SuggestDictionaryImpl {
    constructor(name, source, typosDef) {
        this.name = name;
        this.source = source;
        this.typosDef = typosDef;
        this.containsNoSuggestWords = false;
        this.options = {};
        this.type = 'suggest';
        this.isDictionaryCaseSensitive = true;
        this.size = Object.keys(typosDef).length;
        this.suggestions = (0, util_1.extractAllSuggestions)(typosDef);
        this.suggestionsLower = new Set((0, sync_1.pipe)(this.suggestions, textMappers_1.mapperRemoveCaseAndAccents));
    }
    /**
     * A Forbidden word list does not "have" valid words.
     * Therefore it always returns false.
     * @param _word - the word
     * @param _options - options
     * @returns always false
     */
    has(_word, _options) {
        return false;
    }
    /** A more detailed search for a word, might take longer than `has` */
    find(_word, _options) {
        return undefined;
    }
    isForbidden(_word, _ignoreCaseAndAccents) {
        return false;
    }
    isNoSuggestWord(_word, _options) {
        return false;
    }
    /**
     * Determine if the word can appear in a list of suggestions.
     * @param word - word
     * @param ignoreCaseAndAccents - ignore case.
     * @returns true if a word is suggested, otherwise false.
     */
    isSuggestedWord(word, ignoreCaseAndAccents = defaults.isForbiddenIgnoreCaseAndAccents) {
        if (this.suggestions.has(word))
            return true;
        const lcWord = word.toLowerCase();
        return ignoreCaseAndAccents && (this.suggestions.has(lcWord) || this.suggestionsLower.has(lcWord));
    }
    suggest(word) {
        return this._suggest(word) || this._suggest(word.toLowerCase()) || [];
    }
    _suggest(word) {
        if (!(word in this.typosDef))
            return undefined;
        const sug = this.typosDef[word];
        const isPreferred = true;
        if (!sug)
            return [];
        if (typeof sug === 'string') {
            return [
                {
                    word: sug,
                    cost: 1,
                    isPreferred,
                },
            ];
        }
        return sug.map((word, index) => ({ word, cost: index + 1, isPreferred }));
    }
    genSuggestions(collector) {
        const sugs = this.suggest(collector.word);
        sugs.forEach((result) => collector.add(result));
    }
    mapWord(word) {
        return word;
    }
    getErrors() {
        return [];
    }
}
const createCache = (0, AutoResolve_1.createAutoResolveWeakCache)();
/**
 * Create a dictionary where all words are to be forbidden.
 * @param entries - list of Typos Entries
 * @param name - name of dictionary
 * @param source - source
 * @returns
 */
function createSuggestDictionary(entries, name, source) {
    return createCache.get(entries, () => {
        const def = (0, Typos_1.processEntriesToTyposDef)(entries);
        return new SuggestDictionaryImpl(name, source, def);
    });
}
exports.createSuggestDictionary = createSuggestDictionary;
