// cspell-io/dist/esm/node/file/stat.js
import { promises as fs, statSync } from 'fs';
import { format } from 'util';
import { fetchHead } from './fetch.js';
import { isFileURL, isUrlLike, toURL } from './util.js';
export async function getStat(filenameOrUri) {
    if (isUrlLike(filenameOrUri)) {
        const url = toURL(filenameOrUri);
        if (!isFileURL(url)) {
            try {
                return await getStatHttp(url);
            }
            catch (e) {
                return toError(e);
            }
        }
    }
    return fs.stat(filenameOrUri).catch((e) => toError(e));
}
export function getStatSync(uri) {
    try {
        return statSync(uri);
    }
    catch (e) {
        return toError(e);
    }
}
export async function getStatHttp(url) {
    const headers = await fetchHead(url);
    const eTag = headers.get('etag') || undefined;
    const guessSize = Number.parseInt(headers.get('content-length') || '0', 10);
    return {
        size: eTag ? -1 : guessSize,
        mtimeMs: 0,
        eTag,
    };
}
function toError(e) {
    if (isErrnoException(e) || e instanceof Error)
        return e;
    return new Error(format(e));
}
function isErrnoException(e) {
    if (!e || typeof e !== 'object')
        return false;
    const err = e;
    return err.message !== undefined && err.name !== undefined;
}
