// cspell-lib/dist/esm/util/Uri.mjs
import assert from 'assert';
import { URI, Utils } from 'vscode-uri';
const isFile = /^(?:[a-zA-Z]:|[/\\])/;
const isPossibleUri = /\w:\/\//;
const isUrl = /^(file:|stdin:|https?:|s?ftp:)/;
const STDIN_PROTOCOL = 'stdin:';
export function toUri(uriOrFile) {
    if (UriImpl.isUri(uriOrFile))
        return uriOrFile;
    if (URI.isUri(uriOrFile))
        return UriImpl.from(uriOrFile);
    if (uriOrFile instanceof URL)
        return UriImpl.parse(uriOrFile.toString());
    if (isUrlLike(uriOrFile))
        return UriImpl.parse(uriOrFile.href);
    if (isUri(uriOrFile))
        return UriImpl.from(uriOrFile);
    if (isUrl.test(uriOrFile))
        return UriImpl.parse(uriOrFile);
    return isFile.test(uriOrFile) && !isPossibleUri.test(uriOrFile)
        ? UriImpl.file(normalizeDriveLetter(uriOrFile))
        : UriImpl.parse(uriOrFile);
}
const hasDriveLetter = /^[A-Z]:/i;
export function uriToFilePath(uri) {
    const adj = uri.scheme === 'stdin' ? { scheme: 'file' } : {};
    return normalizeDriveLetter(URI.from(UriImpl.from(uri, adj)).fsPath);
}
export function fromFilePath(file) {
    return UriImpl.file(file);
}
export function fromStdinFilePath(path) {
    return UriImpl.stdin(path);
}
export const file = fromFilePath;
export function parse(uri) {
    return UriImpl.parse(uri);
}
export function normalizeDriveLetter(path) {
    return hasDriveLetter.test(path) ? path[0].toLowerCase() + path.slice(1) : path;
}
function isUrlLike(url) {
    return (!!url && typeof url === 'object' && typeof url.href === 'string') || false;
}
export function isUri(uri) {
    if (!uri || typeof uri !== 'object')
        return false;
    if (UriImpl.isUri(uri))
        return true;
    if (URI.isUri(uri))
        return true;
    const u = uri;
    return typeof u.path === 'string' && typeof u.scheme === 'string';
}
export function basename(uri) {
    return Utils.basename(URI.from(uri));
}
export function dirname(uri) {
    return UriImpl.from(Utils.dirname(URI.from(uri)));
}
export function extname(uri) {
    return Utils.extname(URI.from(uri));
}
export function joinPath(uri, ...paths) {
    return UriImpl.from(Utils.joinPath(URI.from(uri), ...paths));
}
export function resolvePath(uri, ...paths) {
    return UriImpl.from(Utils.resolvePath(URI.from(uri), ...paths));
}
export function from(uri, ...parts) {
    return UriImpl.from(uri, ...parts);
}
const keys = ['scheme', 'authority', 'path', 'query', 'fragment'];
class UriImpl {
    constructor(uri) {
        this.scheme = uri.scheme || '';
        uri.authority && (this.authority = uri.authority);
        this.path = uri.path || '';
        uri.query && (this.query = uri.query);
        uri.fragment && (this.fragment = uri.fragment);
    }
    toString() {
        const path = this.path;
        const base = `${this.scheme}://${this.authority || ''}${path}`;
        const query = (this.query && `?${this.query}`) || '';
        const fragment = (this.fragment && `#${this.fragment}`) || '';
        const url = base + query + fragment;
        return encodeURI(url);
    }
    toJson() {
        const { scheme, authority, path, query, fragment } = this;
        return { scheme, authority, path, query, fragment };
    }
    with(change) {
        const { scheme, authority, path, query, fragment } = this;
        const u = { scheme, authority, path, query, fragment };
        for (const key of keys) {
            if (change[key] && typeof change[key] === 'string') {
                u[key] = change[key];
            }
        }
        return new UriImpl(u);
    }
    static isUri(uri) {
        return uri instanceof UriImpl;
    }
    static from(uri, ...parts) {
        let u = new UriImpl(uri);
        for (const part of parts) {
            u = u.with(part);
        }
        return u;
    }
    static parse(uri) {
        if (uri.startsWith(STDIN_PROTOCOL)) {
            return UriImpl.from(parseStdinUri(uri));
        }
        const u = URI.parse(uri);
        return UriImpl.from(u);
    }
    static file(filename) {
        return UriImpl.from(URI.file(normalizeFilePath(filename)));
    }
    static stdin(filePath = '') {
        return UriImpl.from(UriImpl.file(filePath), { scheme: 'stdin' });
    }
}
function normalizeFilePath(path) {
    return normalizeDriveLetter(path.replace(/\\/g, '/'));
}
function parseStdinUri(uri) {
    assert(uri.startsWith(STDIN_PROTOCOL));
    const idxSlash = STDIN_PROTOCOL.length;
    let idxSlashEnd = idxSlash;
    for (; uri[idxSlashEnd] === '/'; ++idxSlashEnd) {
        // empty
    }
    const pathStart = idxSlashEnd;
    const iH = uri.indexOf('#', pathStart);
    const idxHash = iH > 0 ? iH : uri.length;
    const iQ = uri.indexOf('?', pathStart);
    const idxQ = iQ > 0 && iQ < idxHash ? iQ : idxHash;
    const pathEnd = idxQ;
    const path = uri.slice(pathStart, pathEnd);
    const query = idxQ < idxHash ? uri.slice(idxQ + 1, idxHash) : '';
    const hash = uri.slice(idxHash + 1);
    const pathPrefix = idxSlashEnd - idxSlash > 2 ? '/' : '';
    return {
        scheme: 'stdin',
        path: pathPrefix + normalizeFilePath(decodeURI(path)),
        query: decodeURI(query),
        fragment: decodeURI(hash),
    };
}
export const __testing__ = {
    UriImpl,
    normalizeFilePath,
};
