// cspell-lib/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearCachedFiles = exports.getLanguagesForExt = exports.ExclusionHelper = exports.resolveFile = exports.Link = exports.Text = exports.spellCheckFile = exports.spellCheckDocument = exports.fileToDocument = exports.determineFinalDocumentSettings = exports.constructSettingsForText = exports.combineTextAndLanguageSettings = exports.SpellingDictionaryLoadError = exports.refreshDictionaryCache = exports.isSpellingDictionaryLoadError = exports.getDictionary = exports.createSpellingDictionary = exports.CompoundWordsMethod = exports.defaultSettingsFilename = exports.validateText = exports.IncludeExcludeFlag = exports.checkText = void 0;
__exportStar(require("cspell-io"), exports);
__exportStar(require("./Settings"), exports);
__exportStar(require("@cspell/cspell-types"), exports);
var validator_1 = require("./validator");
Object.defineProperty(exports, "checkText", { enumerable: true, get: function () { return validator_1.checkText; } });
Object.defineProperty(exports, "IncludeExcludeFlag", { enumerable: true, get: function () { return validator_1.IncludeExcludeFlag; } });
Object.defineProperty(exports, "validateText", { enumerable: true, get: function () { return validator_1.validateText; } });
var Settings_1 = require("./Settings");
Object.defineProperty(exports, "defaultSettingsFilename", { enumerable: true, get: function () { return Settings_1.defaultFileName; } });
var SpellingDictionary_1 = require("./SpellingDictionary");
Object.defineProperty(exports, "CompoundWordsMethod", { enumerable: true, get: function () { return SpellingDictionary_1.CompoundWordsMethod; } });
Object.defineProperty(exports, "createSpellingDictionary", { enumerable: true, get: function () { return SpellingDictionary_1.createSpellingDictionary; } });
Object.defineProperty(exports, "getDictionary", { enumerable: true, get: function () { return SpellingDictionary_1.getDictionary; } });
Object.defineProperty(exports, "isSpellingDictionaryLoadError", { enumerable: true, get: function () { return SpellingDictionary_1.isSpellingDictionaryLoadError; } });
Object.defineProperty(exports, "refreshDictionaryCache", { enumerable: true, get: function () { return SpellingDictionary_1.refreshDictionaryCache; } });
Object.defineProperty(exports, "SpellingDictionaryLoadError", { enumerable: true, get: function () { return SpellingDictionary_1.SpellingDictionaryLoadError; } });
var TextDocumentSettings_1 = require("./Settings/TextDocumentSettings");
Object.defineProperty(exports, "combineTextAndLanguageSettings", { enumerable: true, get: function () { return TextDocumentSettings_1.combineTextAndLanguageSettings; } });
var TextDocumentSettings_2 = require("./Settings/TextDocumentSettings");
Object.defineProperty(exports, "constructSettingsForText", { enumerable: true, get: function () { return TextDocumentSettings_2.combineTextAndLanguageSettings; } });
var spellCheckFile_1 = require("./spellCheckFile");
Object.defineProperty(exports, "determineFinalDocumentSettings", { enumerable: true, get: function () { return spellCheckFile_1.determineFinalDocumentSettings; } });
Object.defineProperty(exports, "fileToDocument", { enumerable: true, get: function () { return spellCheckFile_1.fileToDocument; } });
Object.defineProperty(exports, "spellCheckDocument", { enumerable: true, get: function () { return spellCheckFile_1.spellCheckDocument; } });
Object.defineProperty(exports, "spellCheckFile", { enumerable: true, get: function () { return spellCheckFile_1.spellCheckFile; } });
const Text = __importStar(require("./util/text"));
exports.Text = Text;
const Link = __importStar(require("./Settings/link"));
exports.Link = Link;
var resolveFile_1 = require("./util/resolveFile");
Object.defineProperty(exports, "resolveFile", { enumerable: true, get: function () { return resolveFile_1.resolveFile; } });
const ExclusionHelper = __importStar(require("./exclusionHelper"));
exports.ExclusionHelper = ExclusionHelper;
var LanguageIds_1 = require("./LanguageIds");
Object.defineProperty(exports, "getLanguagesForExt", { enumerable: true, get: function () { return LanguageIds_1.getLanguagesForExt; } });
__exportStar(require("./trace"), exports);
const Settings_2 = require("./Settings");
const SpellingDictionary_2 = require("./SpellingDictionary");
async function clearCachedFiles() {
    await Promise.all([Settings_2.clearCachedSettingsFiles(), SpellingDictionary_2.refreshDictionaryCache(0)]);
}
exports.clearCachedFiles = clearCachedFiles;
