// cspell-lib/dist/esm/Settings/Controller/configLoader/configLoader.mjs
import * as json from 'comment-json';
import { cosmiconfig, cosmiconfigSync } from 'cosmiconfig';
import { getDefaultCSpellIO } from 'cspell-io';
import * as path from 'path';
import { createCSpellSettingsInternal as csi } from '../../../Models/CSpellSettingsInternalDef.mjs';
import { AutoResolveLRUCache } from '../../../util/AutoResolveLRUCache.mjs';
import { logError, logWarning } from '../../../util/logger.mjs';
import { resolveFile } from '../../../util/resolveFile.mjs';
import { toUri } from '../../../util/Uri.mjs';
import { configSettingsFileVersion0_1, configSettingsFileVersion0_2, currentSettingsFileVersion, ENV_CSPELL_GLOB_ROOT, } from '../../constants.mjs';
import { mergeSettings } from '../../CSpellSettingsServer.mjs';
import { getRawGlobalSettings } from '../../GlobalSettings.mjs';
import { ImportError } from '../ImportError.mjs';
import { pnpLoader } from '../pnpLoader.mjs';
import { defaultSettings } from './defaultSettings.mjs';
import { normalizeCacheSettings, normalizeDictionaryDefs, normalizeGitignoreRoot, normalizeLanguageSettings, normalizeOverrides, normalizeRawConfig, normalizeReporters, normalizeSettingsGlobs, } from './normalizeRawSettings.mjs';
import { defaultPnPSettings, normalizePnPSettings } from './PnPSettings.mjs';
const supportedCSpellConfigVersions = [configSettingsFileVersion0_2];
const setOfSupportedConfigVersions = Object.freeze(new Set(supportedCSpellConfigVersions));
export const sectionCSpell = 'cSpell';
export const defaultFileName = 'cspell.json';
const gcl = getDefaultConfigLoaderInternal;
const CACHE_SIZE_SEARCH_CONFIG = 32;
/**
 * Logic of the locations:
 * - Support backward compatibility with the VS Code Spell Checker
 *   the spell checker extension can only write to `.json` files because
 *   it would be too difficult to automatically modify a `.js` or `.cjs` file.
 * - To support `cspell.config.js` in a VS Code environment, have a `cspell.json` import
 *   the `cspell.config.js`.
 */
const searchPlaces = Object.freeze([
    'package.json',
    // Original locations
    '.cspell.json',
    'cspell.json',
    '.cSpell.json',
    'cSpell.json',
    // Original locations jsonc
    '.cspell.jsonc',
    'cspell.jsonc',
    // Alternate locations
    '.vscode/cspell.json',
    '.vscode/cSpell.json',
    '.vscode/.cspell.json',
    // Standard Locations
    'cspell.config.json',
    'cspell.config.jsonc',
    'cspell.config.yaml',
    'cspell.config.yml',
    'cspell.yaml',
    'cspell.yml',
    // Dynamic config is looked for last
    'cspell.config.js',
    'cspell.config.cjs',
    // .config
    '.config/.cspell.json',
    '.config/cspell.json',
    '.config/.cSpell.json',
    '.config/cSpell.json',
    '.config/.cspell.jsonc',
    '.config/cspell.jsonc',
    '.config/cspell.config.json',
    '.config/cspell.config.jsonc',
    '.config/cspell.config.yaml',
    '.config/cspell.config.yml',
    '.config/cspell.yaml',
    '.config/cspell.yml',
    '.config/cspell.config.js',
    '.config/cspell.config.cjs',
]);
const cspellCosmiconfig = {
    searchPlaces: searchPlaces.concat(),
    loaders: {
        '.json': parseJson,
        '.jsonc': parseJson,
    },
};
function parseJson(_filename, content) {
    return json.parse(content);
}
export const defaultConfigFilenames = Object.freeze(searchPlaces.concat());
let defaultConfigLoader = undefined;
export class ConfigLoader {
    /**
     * Use `createConfigLoader`
     * @param cspellIO - CSpellIO interface for reading files.
     */
    constructor(cspellIO) {
        this.cspellIO = cspellIO;
        this.cachedFiles = new Map();
        this.cspellConfigExplorer = cosmiconfig('cspell', cspellCosmiconfig);
        this.cspellConfigExplorerSync = cosmiconfigSync('cspell', cspellCosmiconfig);
        this.searchConfigLRU = new AutoResolveLRUCache(CACHE_SIZE_SEARCH_CONFIG, (a, b) => a.searchFrom === b.searchFrom && a.pnpSettings === b.pnpSettings);
    }
    readSettings(filename, relativeToOrDefault, defaultValue) {
        // console.log('Read Settings: %o', { filename, relativeToOrDefault });
        const relativeTo = (typeof relativeToOrDefault === 'string' ? relativeToOrDefault : '') || process.cwd();
        defaultValue = defaultValue || (typeof relativeToOrDefault !== 'string' ? relativeToOrDefault : undefined);
        const ref = resolveFilename(filename, relativeTo);
        return this.importSettings(ref, defaultValue, defaultValue || defaultPnPSettings);
    }
    async readSettingsAsync(filename, relativeTo, pnpSettings) {
        const ref = resolveFilename(filename, relativeTo || process.cwd());
        return this.importSettings(ref, undefined, pnpSettings || defaultPnPSettings);
    }
    /**
     *
     * @param searchFrom the directory / file to start searching from.
     * @param pnpSettings - related to Using Yarn PNP.
     * @returns the resulting settings
     */
    searchForConfig(searchFrom, pnpSettings = defaultPnPSettings) {
        pnpSettings = normalizePnPSettings(pnpSettings);
        return this.searchConfigLRU.get({ searchFrom, pnpSettings }, (p) => this._searchForConfig(p));
    }
    _searchForConfig(params) {
        // console.log('_searchForConfig: %o', { params, stats: this.searchConfigLRU.stats() });
        return gcl()
            .normalizeSearchForConfigResultAsync(params.searchFrom || process.cwd(), this.cspellConfigExplorer.search(params.searchFrom), params.pnpSettings)
            .then((r) => (r.filepath ? r.config : undefined));
    }
    getGlobalSettings() {
        if (!this.globalSettings) {
            const globalConf = getRawGlobalSettings();
            this.globalSettings = {
                id: 'global_config',
                ...this.normalizeSettings(globalConf || {}, './global_config', {}),
            };
        }
        return this.globalSettings;
    }
    clearCachedSettingsFiles() {
        this.searchConfigLRU.clear();
        this.globalSettings = undefined;
        this.cachedFiles.clear();
        this.cspellConfigExplorer.clearCaches();
        this.cspellConfigExplorerSync.clearCaches();
    }
    /**
     * Read a config file and inject the fileRef.
     * @param fileRef - filename plus context, injected into the resulting config.
     */
    readConfig(fileRef) {
        // cspellConfigExplorerSync
        const { filename, error } = fileRef;
        if (error) {
            fileRef.error =
                error instanceof ImportError
                    ? error
                    : new ImportError(`Failed to read config file: "${filename}"`, error);
            return { __importRef: fileRef };
        }
        const s = {};
        try {
            const r = this.cspellConfigExplorerSync.load(filename);
            if (!r?.config)
                throw new Error(`not found: "${filename}"`);
            Object.assign(s, r.config);
            normalizeRawConfig(s);
            validateRawConfig(s, fileRef);
        }
        catch (err) {
            fileRef.error =
                err instanceof ImportError ? err : new ImportError(`Failed to read config file: "${filename}"`, err);
        }
        s.__importRef = fileRef;
        return s;
    }
    importSettings(fileRef, defaultValues, pnpSettings) {
        defaultValues = defaultValues ?? defaultSettings;
        const { filename } = fileRef;
        const importRef = { ...fileRef };
        const cached = this.cachedFiles.get(filename);
        if (cached) {
            const cachedImportRef = cached.__importRef || importRef;
            cachedImportRef.referencedBy = mergeSourceList(cachedImportRef.referencedBy || [], importRef.referencedBy);
            cached.__importRef = cachedImportRef;
            return cached;
        }
        const id = [path.basename(path.dirname(filename)), path.basename(filename)].join('/');
        const name = '';
        const finalizeSettings = csi({ id, name, __importRef: importRef });
        this.cachedFiles.set(filename, finalizeSettings); // add an empty entry to prevent circular references.
        const settings = { ...defaultValues, id, name, ...this.readConfig(importRef) };
        Object.assign(finalizeSettings, this.normalizeSettings(settings, filename, pnpSettings));
        const finalizeSrc = { name: path.basename(filename), ...finalizeSettings.source };
        finalizeSettings.source = { ...finalizeSrc, filename };
        this.cachedFiles.set(filename, finalizeSettings);
        return finalizeSettings;
    }
    /**
     * normalizeSettings handles correcting all relative paths, anchoring globs, and importing other config files.
     * @param rawSettings - raw configuration settings
     * @param pathToSettingsFile - path to the source file of the configuration settings.
     */
    normalizeSettings(rawSettings, pathToSettingsFile, pnpSettings) {
        const id = rawSettings.id ||
            [path.basename(path.dirname(pathToSettingsFile)), path.basename(pathToSettingsFile)].join('/');
        const name = rawSettings.name || id;
        // Try to load any .pnp files before reading dictionaries or other config files.
        const { usePnP = pnpSettings.usePnP, pnpFiles = pnpSettings.pnpFiles } = rawSettings;
        const pnpSettingsToUse = normalizePnPSettings({ usePnP, pnpFiles });
        const pathToSettingsDir = path.dirname(pathToSettingsFile);
        loadPnPSync(pnpSettingsToUse, toUri(pathToSettingsDir));
        // Fix up dictionaryDefinitions
        const settings = {
            version: defaultSettings.version,
            ...rawSettings,
            id,
            name,
            globRoot: resolveGlobRoot(rawSettings, pathToSettingsFile),
            languageSettings: normalizeLanguageSettings(rawSettings.languageSettings),
        };
        const pathToSettings = path.dirname(pathToSettingsFile);
        const normalizedDictionaryDefs = normalizeDictionaryDefs(settings, pathToSettingsFile);
        const normalizedSettingsGlobs = normalizeSettingsGlobs(settings, pathToSettingsFile);
        const normalizedOverrides = normalizeOverrides(settings, pathToSettingsFile);
        const normalizedReporters = normalizeReporters(settings, pathToSettingsFile);
        const normalizedGitignoreRoot = normalizeGitignoreRoot(settings, pathToSettingsFile);
        const normalizedCacheSettings = normalizeCacheSettings(settings, pathToSettingsDir);
        const imports = typeof settings.import === 'string' ? [settings.import] : settings.import || [];
        const source = settings.source || {
            name: settings.name,
            filename: pathToSettingsFile,
        };
        const fileSettings = csi({
            ...settings,
            source,
            ...normalizedDictionaryDefs,
            ...normalizedSettingsGlobs,
            ...normalizedOverrides,
            ...normalizedReporters,
            ...normalizedGitignoreRoot,
            ...normalizedCacheSettings,
        });
        if (!imports.length) {
            return fileSettings;
        }
        const importedSettings = imports
            .map((name) => resolveFilename(name, pathToSettings))
            .map((ref) => ((ref.referencedBy = [source]), ref))
            .map((ref) => this.importSettings(ref, undefined, pnpSettingsToUse))
            .reduce((a, b) => mergeSettings(a, b));
        const finalizeSettings = mergeSettings(importedSettings, fileSettings);
        finalizeSettings.name = settings.name || finalizeSettings.name || '';
        finalizeSettings.id = settings.id || finalizeSettings.id || '';
        return finalizeSettings;
    }
}
class ConfigLoaderInternal extends ConfigLoader {
    constructor(cspellIO) {
        super(cspellIO);
        this._readConfig = this.readConfig.bind(this);
        this._normalizeSettings = this.normalizeSettings.bind(this);
    }
    get _cachedFiles() {
        return this.cachedFiles;
    }
    get _cspellConfigExplorer() {
        return this.cspellConfigExplorer;
    }
    get _cspellConfigExplorerSync() {
        return this.cspellConfigExplorerSync;
    }
    async normalizeSearchForConfigResultAsync(searchPath, searchResult, pnpSettings) {
        let result;
        try {
            result = (await searchResult) || undefined;
        }
        catch (cause) {
            result = new ImportError(`Failed to find config file at: "${searchPath}"`, cause);
        }
        return this.normalizeSearchForConfigResult(searchPath, result, pnpSettings);
    }
    normalizeSearchForConfigResult(searchPath, searchResult, pnpSettings) {
        const error = searchResult instanceof ImportError ? searchResult : undefined;
        const result = searchResult instanceof ImportError ? undefined : searchResult;
        const filepath = result?.filepath;
        if (filepath) {
            const cached = cachedFiles().get(filepath);
            if (cached) {
                return {
                    config: cached,
                    filepath,
                    error,
                };
            }
        }
        const { config = csi({}) } = result || {};
        const filename = result?.filepath ?? searchPath;
        const importRef = { filename: filename, error };
        const id = [path.basename(path.dirname(filename)), path.basename(filename)].join('/');
        const name = result?.filepath ? id : `Config not found: ${id}`;
        const finalizeSettings = csi({ id, name, __importRef: importRef });
        const settings = { id, ...config };
        cachedFiles().set(filename, finalizeSettings); // add an empty entry to prevent circular references.
        Object.assign(finalizeSettings, this.normalizeSettings(settings, filename, pnpSettings));
        return {
            config: finalizeSettings,
            filepath,
            error,
        };
    }
}
function mergeSourceList(orig, append) {
    const collection = new Map(orig.map((s) => [s.name + (s.filename || ''), s]));
    for (const s of append || []) {
        const key = s.name + (s.filename || '');
        if (!collection.has(key)) {
            collection.set(key, s);
        }
    }
    return [...collection.values()];
}
/**
 *
 * @param searchFrom the directory / file to start searching from.
 * @param pnpSettings - related to Using Yarn PNP.
 * @returns the resulting settings
 */
export function searchForConfig(searchFrom, pnpSettings = defaultPnPSettings) {
    return gcl().searchForConfig(searchFrom, pnpSettings);
}
/**
 *
 * @param searchFrom the directory / file to start searching from.
 * @param pnpSettings - related to Using Yarn PNP.
 * @returns the resulting settings
 * @deprecated
 * @deprecationMessage Use `searchForConfig`. It is very difficult to support Sync files when settings include web requests.
 */
export function searchForConfigSync(searchFrom, pnpSettings = defaultPnPSettings) {
    pnpSettings = normalizePnPSettings(pnpSettings);
    let searchResult;
    try {
        searchResult = cspellConfigExplorerSync().search(searchFrom) || undefined;
    }
    catch (err) {
        searchResult = new ImportError(`Failed to find config file from: "${searchFrom}"`, err);
    }
    return gcl().normalizeSearchForConfigResult(searchFrom || process.cwd(), searchResult, pnpSettings).config;
}
/**
 * Load a CSpell configuration files.
 * @param file - path or package reference to load.
 * @param pnpSettings - PnP settings
 * @returns normalized CSpellSettings
 */
export async function loadConfig(file, pnpSettings = defaultPnPSettings) {
    return gcl().readSettingsAsync(file, undefined, pnpSettings);
}
/**
 * Load a CSpell configuration files.
 * @param filename - path or package reference to load.
 * @param pnpSettings - PnP settings
 * @returns normalized CSpellSettings
 * @deprecated
 */
export function loadConfigSync(filename, pnpSettings = defaultPnPSettings) {
    const pnp = normalizePnPSettings(pnpSettings);
    return gcl().readSettings(filename, pnp);
}
export function loadPnP(pnpSettings, searchFrom) {
    if (!pnpSettings.usePnP) {
        return Promise.resolve(undefined);
    }
    const loader = pnpLoader(pnpSettings.pnpFiles);
    return loader.load(searchFrom);
}
export function loadPnPSync(pnpSettings, searchFrom) {
    if (!pnpSettings.usePnP) {
        return undefined;
    }
    const loader = pnpLoader(pnpSettings.pnpFiles);
    return loader.loadSync(searchFrom);
}
export function readRawSettings(filename, relativeTo) {
    relativeTo = relativeTo || process.cwd();
    const ref = resolveFilename(filename, relativeTo);
    return gcl()._readConfig(ref);
}
function resolveFilename(filename, relativeTo) {
    const r = resolveFile(filename, relativeTo);
    return {
        filename: r.filename,
        error: r.found ? undefined : new Error(`Failed to resolve file: "${filename}"`),
    };
}
export function getGlobalSettings() {
    return gcl().getGlobalSettings();
}
export function getCachedFileSize() {
    return cachedFiles().size;
}
export function clearCachedSettingsFiles() {
    return gcl().clearCachedSettingsFiles();
}
const nestedConfigDirectories = {
    '.vscode': true,
    '.config': true,
};
function resolveGlobRoot(settings, pathToSettingsFile) {
    const settingsFileDirRaw = path.dirname(pathToSettingsFile);
    const settingsFileDirName = path.basename(settingsFileDirRaw);
    const isNestedConfig = settingsFileDirName in nestedConfigDirectories;
    const isVSCode = settingsFileDirName === '.vscode';
    const settingsFileDir = isNestedConfig ? path.dirname(settingsFileDirRaw) : settingsFileDirRaw;
    const envGlobRoot = process.env[ENV_CSPELL_GLOB_ROOT];
    const defaultGlobRoot = envGlobRoot ?? '${cwd}';
    const rawRoot = settings.globRoot ??
        (settings.version === configSettingsFileVersion0_1 ||
            (envGlobRoot && !settings.version) ||
            (isVSCode && !settings.version)
            ? defaultGlobRoot
            : settingsFileDir);
    const globRoot = rawRoot.startsWith('${cwd}') ? rawRoot : path.resolve(settingsFileDir, rawRoot);
    return globRoot;
}
function validationMessage(msg, fileRef) {
    return msg + `\n  File: "${fileRef.filename}"`;
}
function validateRawConfigVersion(config, fileRef) {
    const { version } = config;
    if (version === undefined)
        return;
    if (typeof version !== 'string') {
        logError(validationMessage(`Unsupported config file version: "${version}", string expected`, fileRef));
        return;
    }
    if (setOfSupportedConfigVersions.has(version))
        return;
    if (!/^\d+(\.\d+)*$/.test(version)) {
        logError(validationMessage(`Unsupported config file version: "${version}"`, fileRef));
        return;
    }
    const msg = version > currentSettingsFileVersion
        ? `Newer config file version found: "${version}". Supported version is "${currentSettingsFileVersion}"`
        : `Legacy config file version found: "${version}", upgrade to "${currentSettingsFileVersion}"`;
    logWarning(validationMessage(msg, fileRef));
}
function validateRawConfigExports(config, fileRef) {
    if (config.default) {
        throw new ImportError(validationMessage('Module `export default` is not supported.\n  Use `module.exports =` instead.', fileRef));
    }
}
function validateRawConfig(config, fileRef) {
    const validations = [validateRawConfigExports, validateRawConfigVersion];
    validations.forEach((fn) => fn(config, fileRef));
}
function createConfigLoaderInternal(cspellIO) {
    return new ConfigLoaderInternal(cspellIO ?? getDefaultCSpellIO());
}
export function createConfigLoader(cspellIO) {
    return createConfigLoaderInternal(cspellIO);
}
function getDefaultConfigLoaderInternal() {
    if (defaultConfigLoader)
        return defaultConfigLoader;
    return (defaultConfigLoader = createConfigLoaderInternal());
}
export function getDefaultConfigLoader() {
    return getDefaultConfigLoaderInternal();
}
function cachedFiles() {
    return gcl()._cachedFiles;
}
function cspellConfigExplorerSync() {
    return gcl()._cspellConfigExplorerSync;
}
export const __testing__ = {
    getDefaultConfigLoaderInternal,
    normalizeCacheSettings,
    validateRawConfigExports,
    validateRawConfigVersion,
};
