// cspell-grammar/dist/parser/grammarTypesHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isGrammar = exports.isPatternPatterns = exports.isPatternBeginEnd = exports.isPatternMatch = exports.isPatternInclude = void 0;
function isPatternInclude(p) {
    return !!p.include;
}
exports.isPatternInclude = isPatternInclude;
const TypeofMatch = {
    object: true,
    string: true,
};
const TypeofBegin = TypeofMatch;
const TypeofEnd = { ...TypeofBegin, undefined: true };
function isPatternMatch(pattern) {
    const p = pattern;
    return !!p.match && typeof p.match in TypeofMatch;
}
exports.isPatternMatch = isPatternMatch;
function isPatternBeginEnd(pattern) {
    const p = pattern;
    return p.begin !== undefined && typeof p.begin in TypeofBegin && typeof p.end in TypeofEnd;
}
exports.isPatternBeginEnd = isPatternBeginEnd;
function isPatternPatterns(p) {
    return Array.isArray(p.patterns);
}
exports.isPatternPatterns = isPatternPatterns;
function isGrammar(g) {
    return g.scopeName !== undefined;
}
exports.isGrammar = isGrammar;
