// cspell-trie-lib/dist/lib/mappers/mapDictionaryInfoToWeightMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__testing__ = exports.hunspellInformationToSuggestionCostDef = exports.dictionaryInformationToWeightMap = void 0;
const weightedMaps_1 = require("../distance/weightedMaps");
const util_1 = require("../utils/util");
function dictionaryInformationToWeightMap(dictInfo) {
    const defsEC = dictInfo.suggestionEditCosts || [];
    const defsHI = dictInfo.hunspellInformation
        ? hunspellInformationToSuggestionCostDef(dictInfo.hunspellInformation)
        : [];
    return (0, weightedMaps_1.createWeightMap)(...defsEC, ...defsHI);
}
exports.dictionaryInformationToWeightMap = dictionaryInformationToWeightMap;
function hunspellInformationToSuggestionCostDef(hunInfo) {
    const defs = [];
    const costs = calcCosts(hunInfo.costs);
    function parseAff(aff, costs) {
        // cspell:ignore OCONV
        const regSupportedAff = /^(?:MAP|KEY|TRY|NO-TRY|ICONV|OCONV|REP)\s/;
        const rejectAff = /^(?:MAP|KEY|TRY|ICONV|OCONV|REP)\s+\d+$/;
        const lines = aff
            .split('\n')
            .map((a) => a.replace(/#.*/, ''))
            .map((a) => a.trim())
            .filter((a) => regSupportedAff.test(a))
            .filter((a) => !rejectAff.test(a));
        lines.forEach((line) => {
            [affMap, affNoTry, affRepConv, affTry, affTryFirstCharacterReplace, affKey]
                .map((fn) => fn(line, costs))
                .filter(util_1.isDefined)
                .forEach((def) => defs.push(def));
        });
    }
    parseAff(hunInfo.aff, costs);
    return defs;
}
exports.hunspellInformationToSuggestionCostDef = hunspellInformationToSuggestionCostDef;
function calcCosts(costs = {}) {
    const { firstLetterPenalty = 4, ioConvertCost = 30, keyboardCost = 94, mapCost = 25, replaceCosts = 75, tryCharCost = 95, } = costs;
    const c = { tryCharCost, keyboardCost, mapCost, ioConvertCost, replaceCosts, firstLetterPenalty };
    return c;
}
const regExpMap = /^(?:MAP)\s+(\S+)$/;
function affMap(line, costs) {
    const m = line.match(regExpMap);
    if (!m)
        return undefined;
    const map = m[1];
    const cost = costs.mapCost;
    return {
        map,
        replace: cost,
        swap: cost,
    };
}
const regExpTry = /^(?:TRY)\s+(\S+)$/;
function affTry(line, costs) {
    const m = line.match(regExpTry);
    if (!m)
        return undefined;
    const map = m[1];
    const cost = costs.tryCharCost;
    return {
        map,
        insDel: cost,
        replace: cost,
        swap: cost,
    };
}
function affTryFirstCharacterReplace(line, costs) {
    const m = line.match(regExpTry);
    if (!m)
        return undefined;
    const map = [...split(m[1])].map((char) => `(^${char})`).join('');
    const cost = costs.tryCharCost;
    const penalty = costs.firstLetterPenalty;
    return {
        map,
        replace: cost - penalty,
        penalty,
    };
}
const regExpNoTry = /^NO-TRY\s+(\S+)$/;
function affNoTry(line, costs) {
    const m = line.match(regExpNoTry);
    if (!m)
        return undefined;
    const map = m[1];
    return {
        map,
        insDel: Math.max(100 - costs.tryCharCost, 0),
        penalty: 100 + costs.tryCharCost,
    };
}
// cspell:ignore conv
const regExpRepConv = /^(?:REP|(?:I|O)CONV)\s+(\S+)\s+(\S+)$/;
function affRepConv(line, costs) {
    const m = line.match(regExpRepConv);
    if (!m)
        return undefined;
    const cost = line.startsWith('REP') ? costs.replaceCosts : costs.ioConvertCost;
    const from = m[1];
    let into = m[2];
    into = into.replace(/^0$/, '');
    if (from.startsWith('^') && !into.startsWith('^')) {
        into = '^' + into;
    }
    if (from.endsWith('$') && !into.endsWith('$')) {
        into = into + '$';
    }
    return {
        map: `(${from})(${into})`,
        replace: cost,
    };
}
const regExpKey = /^(?:KEY)\s+(\S+)$/;
function affKey(line, costs) {
    const m = line.match(regExpKey);
    if (!m)
        return undefined;
    const kbd = m[1];
    const pairs = [...split(kbd)]
        .map(reducer((p, v) => ({ a: p.b, b: v }), { a: '|', b: '|' }))
        .filter((ab) => ab.a !== '|' && ab.b !== '|')
        .map((ab) => ab.a + ab.b);
    const map = pairs.join('|');
    const cost = costs.keyboardCost;
    return {
        map,
        replace: cost,
        swap: cost,
    };
}
function* split(map) {
    let seq = '';
    let mode = 0;
    for (const char of map) {
        if (mode && char === ')') {
            yield seq;
            mode = 0;
            continue;
        }
        if (mode) {
            seq += char;
            continue;
        }
        if (char === '(') {
            mode = 1;
            seq = '';
            continue;
        }
        yield char;
    }
}
function reducer(fn, initialVal) {
    let acc = initialVal;
    return (val, i) => (acc = fn(acc, val, i));
}
// function pipe<T>(v: T, ...fns: ((v: T) => T)[]): T {
//     for (const fn of fns) {
//         v = fn(v);
//     }
//     return v;
// }
exports.__testing__ = {
    affKey,
    affMap,
    affNoTry,
    affRepConv,
    affTry,
    affTryFirstCharacterReplace,
    split,
    calcCosts,
};
