// cspell-lib/dist/cjs/Settings/GlobalSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGlobalConfigPath = exports.writeRawGlobalSettings = exports.getRawGlobalSettings = void 0;
const util_1 = require("util");
const errors_js_1 = require("../util/errors.js");
const logger_js_1 = require("../util/logger.js");
const cfgStore_js_1 = require("./cfgStore.js");
const packageName = 'cspell';
function getRawGlobalSettings() {
    const name = 'CSpell Configstore';
    const globalConf = {
        source: {
            name,
            filename: undefined,
        },
    };
    try {
        // console.warn('%o', ConfigStore);
        const cfgStore = new cfgStore_js_1.ConfigStore(packageName);
        // console.warn('%o', cfgStore);
        const cfg = cfgStore.all;
        // Only populate globalConf is there are values.
        if (cfg && Object.keys(cfg).length) {
            Object.assign(globalConf, cfg);
            globalConf.source = {
                name,
                filename: cfgStore.path,
            };
        }
    }
    catch (error) {
        if (!(0, errors_js_1.isErrnoException)(error) ||
            !error.code ||
            !['ENOENT', 'EACCES', 'ENOTDIR', 'EISDIR'].includes(error.code)) {
            (0, logger_js_1.logError)(error);
        }
    }
    return globalConf;
}
exports.getRawGlobalSettings = getRawGlobalSettings;
function writeRawGlobalSettings(settings) {
    const toWrite = {
        import: settings.import,
    };
    try {
        const cfgStore = new cfgStore_js_1.ConfigStore(packageName);
        cfgStore.set(toWrite);
        return undefined;
    }
    catch (error) {
        if (error instanceof Error)
            return error;
        return new Error((0, util_1.format)(error));
    }
}
exports.writeRawGlobalSettings = writeRawGlobalSettings;
function getGlobalConfigPath() {
    const cfgStore = new cfgStore_js_1.ConfigStore(packageName);
    return cfgStore.path;
}
exports.getGlobalConfigPath = getGlobalConfigPath;
