// cspell-lib/dist/cjs/Settings/LanguageSettings.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.calcSettingsForLanguageId = exports.calcUserSettingsForLanguage = exports.calcSettingsForLanguage = exports.isValidLocaleIntlFormat = exports.isLocaleInSet = exports.normalizeLocaleIntl = exports.normalizeLocale = exports.normalizeLanguageId = exports.getDefaultLanguageSettings = void 0;
const AutoResolve_js_1 = require("../util/AutoResolve.js");
const util_js_1 = require("../util/util.js");
const SpellSettings = __importStar(require("./CSpellSettingsServer.js"));
const defaultLocale = 'en';
const defaultLanguageSettings = [];
function getDefaultLanguageSettings() {
    return defaultLanguageSettings;
}
exports.getDefaultLanguageSettings = getDefaultLanguageSettings;
function localesToList(locales) {
    return stringToList(locales.replace(/\s+/g, ','));
}
function stringToList(sList) {
    return sList
        .replace(/[|;]/g, ',')
        .split(',')
        .map((s) => s.trim())
        .filter((s) => !!s);
}
function memorizer(resolver) {
    const cache = (0, AutoResolve_js_1.createAutoResolveCache)();
    return (k) => cache.get(k, resolver);
}
const _normalizeLanguageId = memorizer(__normalizeLanguageId);
function __normalizeLanguageId(langId) {
    const langIds = stringToList(langId);
    return new Set(langIds.map((a) => a.toLowerCase()));
}
function normalizeLanguageId(langId) {
    return _normalizeLanguageId(typeof langId === 'string' ? langId : langId.join(','));
}
exports.normalizeLanguageId = normalizeLanguageId;
const _normalizeLocale = memorizer(__normalizeLocale);
function __normalizeLocale(locale) {
    const locales = localesToList(locale);
    return new Set(locales.map((locale) => locale.toLowerCase().replace(/[^a-z]/g, '')));
}
function normalizeLocale(locale) {
    locale = typeof locale === 'string' ? locale : locale.join(',');
    return _normalizeLocale(locale);
}
exports.normalizeLocale = normalizeLocale;
function normalizeLocaleIntl(locale) {
    const values = [...normalizeLocale(locale)].map((locale) => locale.replace(/^([a-z]{2})-?([a-z]{2})$/, (_, lang, locale) => locale ? `${lang}-${locale.toUpperCase()}` : lang));
    return new Set(values);
}
exports.normalizeLocaleIntl = normalizeLocaleIntl;
function isLocaleInSet(locale, setOfLocals) {
    const locales = normalizeLocale(locale);
    return (0, util_js_1.doSetsIntersect)(locales, setOfLocals);
}
exports.isLocaleInSet = isLocaleInSet;
const regExpValidIntlLocaleStrict = /^[a-z]{2}(-[A-Z]{2})?$/;
const regExpValidIntlLocale = new RegExp(regExpValidIntlLocaleStrict, 'i');
/**
 * Test if a locale should be ok with Intl
 * @param locale - locale string
 * @param strict - case must match
 * @returns true if it matches the standard 2 letter or 4 letter forms.
 */
function isValidLocaleIntlFormat(locale, strict = false) {
    if (typeof locale === 'string')
        return strict ? regExpValidIntlLocaleStrict.test(locale) : regExpValidIntlLocale.test(locale);
    for (const item of locale) {
        if (!isValidLocaleIntlFormat(item, strict))
            return false;
    }
    return locale.length > 0;
}
exports.isValidLocaleIntlFormat = isValidLocaleIntlFormat;
const cacheCalcSettingsForLanguage = new WeakMap();
function calcSettingsForLanguage(languageSettings, languageId, locale) {
    const mapLang = (0, AutoResolve_js_1.autoResolveWeak)(cacheCalcSettingsForLanguage, languageSettings, () => new Map());
    const mapLocale = (0, AutoResolve_js_1.autoResolve)(mapLang, languageId, () => new Map());
    return (0, AutoResolve_js_1.autoResolve)(mapLocale, locale, () => _calcSettingsForLanguage(languageSettings, languageId, locale));
}
exports.calcSettingsForLanguage = calcSettingsForLanguage;
function _calcSettingsForLanguage(languageSettings, languageId, locale) {
    languageId = languageId.toLowerCase();
    const allowedLocals = normalizeLocale(locale);
    const ls = languageSettings
        .filter((s) => doesLanguageSettingMatchLanguageId(s, languageId))
        .filter((s) => !s.locale || s.locale === '*' || isLocaleInSet(s.locale, allowedLocals))
        .map((langSetting) => {
        const { languageId: _languageId, locale: _locale, ...s } = langSetting;
        return s;
    })
        .reduce((langSetting, setting) => SpellSettings.mergeSettings(langSetting, setting), {});
    ls.languageId = languageId;
    ls.locale = locale;
    return ls;
}
const cacheDoesLanguageSettingMatchLanguageId = new WeakMap();
function doesLanguageSettingMatchLanguageId(s, languageId) {
    const r = cacheDoesLanguageSettingMatchLanguageId.get(s) ?? new Map();
    const f = r.get(languageId);
    if (f !== undefined) {
        return f;
    }
    const v = _doesLanguageSettingMatchLanguageId(s, languageId);
    r.set(languageId, v);
    cacheDoesLanguageSettingMatchLanguageId.set(s, r);
    return v;
}
function _doesLanguageSettingMatchLanguageId(s, languageId) {
    const languageSettingsLanguageIds = s.languageId;
    if (!languageSettingsLanguageIds || languageSettingsLanguageIds === '*')
        return true;
    const ids = normalizeLanguageId(languageSettingsLanguageIds);
    if (ids.has(languageId))
        return true;
    if (ids.has('!' + languageId))
        return false;
    const numExcludes = [...ids].filter((id) => id.startsWith('!')).length;
    return numExcludes === ids.size;
}
function calcUserSettingsForLanguage(settings, languageId) {
    const { languageSettings = [], language: locale = defaultLocale, allowCompoundWords, enabled } = settings;
    const langSettings = {
        allowCompoundWords,
        enabled,
        ...calcSettingsForLanguage(languageSettings, languageId, locale),
    };
    return SpellSettings.mergeSettings(settings, langSettings);
}
exports.calcUserSettingsForLanguage = calcUserSettingsForLanguage;
function calcSettingsForLanguageId(baseSettings, languageId) {
    const langIds = ['*', ...normalizeLanguageId(languageId)];
    const langSettings = langIds.reduce((settings, languageId) => {
        return calcUserSettingsForLanguage(settings, languageId);
    }, baseSettings);
    return langSettings;
}
exports.calcSettingsForLanguageId = calcSettingsForLanguageId;
