// cspell/dist/CSpellApplicationConfiguration.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CSpellApplicationConfiguration = exports.defaultConfigGlobOptions = void 0;
const glob_1 = require("./util/glob");
const path = __importStar(require("path"));
const util = __importStar(require("./util/util"));
const emitters_1 = require("./emitters");
const defaultContextRange = 20;
const nullEmitter = () => {
    /* empty */
};
// cspell:word nocase
const defaultMinimatchOptions = { nocase: true };
exports.defaultConfigGlobOptions = defaultMinimatchOptions;
class CSpellApplicationConfiguration {
    constructor(files, options, emitters) {
        this.files = files;
        this.options = options;
        this.emitters = emitters;
        this.configGlobOptions = exports.defaultConfigGlobOptions;
        this.root = path.resolve(options.root || process.cwd());
        this.info = emitters.info || nullEmitter;
        this.debug = emitters.debug || ((msg) => this.info(msg, emitters_1.MessageTypes.Debug));
        this.configFile = options.config;
        this.excludes = glob_1.calcExcludeGlobInfo(this.root, options.exclude);
        this.logIssue = emitters.issue || nullEmitter;
        this.locale = options.locale || options.local || '';
        this.uniqueFilter = options.unique ? util.uniqueFilterFnGenerator((issue) => issue.text) : () => true;
        this.progress = emitters.progress || nullEmitter;
        this.showContext =
            options.showContext === true ? defaultContextRange : options.showContext ? options.showContext : 0;
    }
}
exports.CSpellApplicationConfiguration = CSpellApplicationConfiguration;
