// cspell-io/dist/file/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStatSync = exports.getStat = exports.writeToFileIterableP = exports.writeToFileIterable = exports.writeToFile = exports.readFileSync = exports.readFile = void 0;
var file_1 = require("./../node/file");
Object.defineProperty(exports, "readFile", { enumerable: true, get: function () { return file_1.readFile; } });
Object.defineProperty(exports, "readFileSync", { enumerable: true, get: function () { return file_1.readFileSync; } });
var file_2 = require("./../node/file");
Object.defineProperty(exports, "writeToFile", { enumerable: true, get: function () { return file_2.writeToFile; } });
Object.defineProperty(exports, "writeToFileIterable", { enumerable: true, get: function () { return file_2.writeToFileIterable; } });
Object.defineProperty(exports, "writeToFileIterableP", { enumerable: true, get: function () { return file_2.writeToFileIterableP; } });
var file_3 = require("./../node/file");
Object.defineProperty(exports, "getStat", { enumerable: true, get: function () { return file_3.getStat; } });
Object.defineProperty(exports, "getStatSync", { enumerable: true, get: function () { return file_3.getStatSync; } });
