// cspell-io/dist/CSpellIOWeb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CSpellIOWeb = void 0;
const ErrorNotImplemented_1 = require("./errors/ErrorNotImplemented");
const stat_1 = require("./common/stat");
class CSpellIOWeb {
    readFile(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('readFile');
    }
    readFileSync(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('readFileSync');
    }
    writeFile(_uriOrFilename, _content) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('writeFile');
    }
    getStat(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('getStat');
    }
    getStatSync(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('getStatSync');
    }
    compareStats(left, right) {
        return (0, stat_1.compareStats)(left, right);
    }
    toURL(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('toURL');
    }
    uriBasename(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('uriBasename');
    }
    uriDirname(_uriOrFilename) {
        throw new ErrorNotImplemented_1.ErrorNotImplemented('uriDirname');
    }
}
exports.CSpellIOWeb = CSpellIOWeb;
