// cspell-lib/dist/SpellingDictionary/createSpellingDictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFailedToLoadDictionary = exports.createSpellingDictionary = void 0;
const cspell_trie_lib_1 = require("cspell-trie-lib");
const SpellingDictionaryFromTrie_1 = require("./SpellingDictionaryFromTrie");
function createSpellingDictionary(wordList, name, source, options) {
    // console.log(`createSpellingDictionary ${name} ${source}`);
    const words = cspell_trie_lib_1.parseDictionaryLines(wordList);
    const trie = cspell_trie_lib_1.buildTrieFast(words);
    return new SpellingDictionaryFromTrie_1.SpellingDictionaryFromTrie(trie, name, options, source);
}
exports.createSpellingDictionary = createSpellingDictionary;
function createFailedToLoadDictionary(error) {
    const { options, uri: source } = error;
    const errors = [error];
    return {
        name: options.name,
        source,
        type: 'error',
        has: () => false,
        isForbidden: () => false,
        suggest: () => [],
        mapWord: (a) => a,
        genSuggestions: () => {
            return;
        },
        size: 0,
        options,
        isDictionaryCaseSensitive: false,
        getErrors: () => errors,
    };
}
exports.createFailedToLoadDictionary = createFailedToLoadDictionary;
