// cspell-lib/dist/SpellingDictionary/createSpellingDictionary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFailedToLoadDictionary = exports.createForbiddenWordsDictionary = exports.createSpellingDictionary = void 0;
const cspell_trie_lib_1 = require("cspell-trie-lib");
const SpellingDictionaryFromTrie_1 = require("./SpellingDictionaryFromTrie");
const gensequence_1 = require("gensequence");
function createSpellingDictionary(wordList, name, source, options) {
    // console.log(`createSpellingDictionary ${name} ${source}`);
    const words = cspell_trie_lib_1.parseDictionaryLines(wordList);
    const trie = cspell_trie_lib_1.buildTrieFast(words);
    return new SpellingDictionaryFromTrie_1.SpellingDictionaryFromTrie(trie, name, options || {}, source);
}
exports.createSpellingDictionary = createSpellingDictionary;
function createForbiddenWordsDictionary(wordList, name, source, options) {
    // console.log(`createSpellingDictionary ${name} ${source}`);
    const words = cspell_trie_lib_1.parseDictionaryLines(wordList.concat(wordList.map((a) => a.toLowerCase())), {
        stripCaseAndAccents: !(options === null || options === void 0 ? void 0 : options.noSuggest),
    });
    const forbidWords = gensequence_1.operators.map((w) => '!' + w)(words);
    const trie = cspell_trie_lib_1.buildTrieFast(forbidWords);
    return new SpellingDictionaryFromTrie_1.SpellingDictionaryFromTrie(trie, name, options || {}, source);
}
exports.createForbiddenWordsDictionary = createForbiddenWordsDictionary;
function createFailedToLoadDictionary(error) {
    const { options, uri: source } = error;
    const errors = [error];
    return {
        name: options.name,
        source,
        type: 'error',
        containsNoSuggestWords: false,
        has: () => false,
        find: () => undefined,
        isNoSuggestWord: () => false,
        isForbidden: () => false,
        suggest: () => [],
        mapWord: (a) => a,
        genSuggestions: () => {
            return;
        },
        size: 0,
        options,
        isDictionaryCaseSensitive: false,
        getErrors: () => errors,
    };
}
exports.createFailedToLoadDictionary = createFailedToLoadDictionary;
