// cspell-lib/dist/SpellingDictionary/Dictionaries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDictionaryInternal = exports.refreshDictionaryCache = exports.loadDictionaryDefs = void 0;
const DictionarySettings_1 = require("../Settings/DictionarySettings");
const util_1 = require("../util/util");
const createSpellingDictionary_1 = require("./createSpellingDictionary");
const DictionaryLoader_1 = require("./DictionaryLoader");
const SpellingDictionaryCollection_1 = require("./SpellingDictionaryCollection");
function loadDictionaryDefs(defsToLoad) {
    return defsToLoad.map((def) => (0, DictionaryLoader_1.loadDictionary)(def.path, def));
}
exports.loadDictionaryDefs = loadDictionaryDefs;
function refreshDictionaryCache(maxAge) {
    return (0, DictionaryLoader_1.refreshCacheEntries)(maxAge);
}
exports.refreshDictionaryCache = refreshDictionaryCache;
const emptyWords = Object.freeze([]);
function getDictionaryInternal(settings) {
    const { words = emptyWords, userWords = emptyWords, flagWords = emptyWords, ignoreWords = emptyWords } = settings;
    const spellDictionaries = loadDictionaryDefs((0, DictionarySettings_1.calcDictionaryDefsToLoad)(settings));
    const settingsWordsDictionary = (0, createSpellingDictionary_1.createSpellingDictionary)(words, '[words]', 'From Settings `words`', {
        caseSensitive: true,
        weightMap: undefined,
    });
    const settingsUserWordsDictionary = userWords.length
        ? (0, createSpellingDictionary_1.createSpellingDictionary)(userWords, '[userWords]', 'From Settings `userWords`', {
            caseSensitive: true,
            weightMap: undefined,
        })
        : undefined;
    const ignoreWordsDictionary = (0, createSpellingDictionary_1.createSpellingDictionary)(ignoreWords, '[ignoreWords]', 'From Settings `ignoreWords`', {
        caseSensitive: true,
        noSuggest: true,
        weightMap: undefined,
    });
    const flagWordsDictionary = (0, createSpellingDictionary_1.createForbiddenWordsDictionary)(flagWords, '[flagWords]', 'From Settings `flagWords`', {
        weightMap: undefined,
    });
    const dictionaries = [
        ...spellDictionaries,
        settingsWordsDictionary,
        settingsUserWordsDictionary,
        ignoreWordsDictionary,
        flagWordsDictionary,
    ].filter(util_1.isDefined);
    return (0, SpellingDictionaryCollection_1.createCollectionP)(dictionaries, 'dictionary collection');
}
exports.getDictionaryInternal = getDictionaryInternal;
