// cspell-trie-lib/dist/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseDictionary = exports.parseDictionaryLines = exports.suggestionCollector = exports.TrieBuilder = exports.buildTrieFast = exports.buildTrie = exports.ChildMap = exports.FLAG_WORD = void 0;
__exportStar(require("./trie"), exports);
var TrieNode_1 = require("./TrieNode");
Object.defineProperty(exports, "FLAG_WORD", { enumerable: true, get: function () { return TrieNode_1.FLAG_WORD; } });
Object.defineProperty(exports, "ChildMap", { enumerable: true, get: function () { return TrieNode_1.ChildMap; } });
__exportStar(require("./util"), exports);
__exportStar(require("./walker"), exports);
__exportStar(require("./importExport"), exports);
var TrieBuilder_1 = require("./TrieBuilder");
Object.defineProperty(exports, "buildTrie", { enumerable: true, get: function () { return TrieBuilder_1.buildTrie; } });
Object.defineProperty(exports, "buildTrieFast", { enumerable: true, get: function () { return TrieBuilder_1.buildTrieFast; } });
Object.defineProperty(exports, "TrieBuilder", { enumerable: true, get: function () { return TrieBuilder_1.TrieBuilder; } });
__exportStar(require("./consolidate"), exports);
var suggestCollector_1 = require("./suggestCollector");
Object.defineProperty(exports, "suggestionCollector", { enumerable: true, get: function () { return suggestCollector_1.suggestionCollector; } });
var SimpleDictionaryParser_1 = require("./SimpleDictionaryParser");
Object.defineProperty(exports, "parseDictionaryLines", { enumerable: true, get: function () { return SimpleDictionaryParser_1.parseDictionaryLines; } });
Object.defineProperty(exports, "parseDictionary", { enumerable: true, get: function () { return SimpleDictionaryParser_1.parseDictionary; } });
