// cspell-lib/dist/util/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__testing__ = exports.isError = exports.isErrnoException = void 0;
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function getTypeOf(t) {
    return typeof t;
}
const allowStringOrUndefined = {
    string: true,
    undefined: true,
};
const allowNumberOrUndefined = {
    number: true,
    undefined: true,
};
function isErrnoException(e) {
    if (!e || typeof e !== 'object')
        return false;
    if (!isError(e))
        return false;
    const ex = e;
    return (typeof ex.errno in allowNumberOrUndefined &&
        typeof ex.code in allowStringOrUndefined &&
        typeof ex.path in allowStringOrUndefined);
}
exports.isErrnoException = isErrnoException;
function isError(e) {
    if (e instanceof Error)
        return true;
    if (!e || typeof e !== 'object')
        return false;
    const ex = e;
    return typeof ex.name == 'string' && typeof ex.message == 'string' && typeof ex.stack in allowStringOrUndefined;
}
exports.isError = isError;
exports.__testing__ = {
    getTypeOf,
};
