// cspell-lib/dist/util/textRegex.js
"use strict";
// cspell:ignore ings ning gimuy anrvtbf
Object.defineProperty(exports, "__esModule", { value: true });
exports.regExTrailingEndings = exports.regExDanglingQuote = exports.regExEscapeCharacters = exports.regExAccents = exports.regExMatchRegExParts = exports.regExPossibleWordBreaks = exports.regExAllLower = exports.regExAllUpper = exports.regExFirstUpper = exports.regExIgnoreCharacters = exports.regExWordsAndDigits = exports.regExWords = exports.regExSplitWords2 = exports.regExSplitWords = exports.regExUpperSOrIng = exports.regExLines = void 0;
exports.regExLines = /.*(\r?\n|$)/g;
exports.regExUpperSOrIng = /(\p{Lu}+\\?['’]?(?:s|ing|ies|es|ings|ed|ning))(?!\p{Ll})/gu;
exports.regExSplitWords = /(\p{Ll})(\p{Lu})/gu;
exports.regExSplitWords2 = /(\p{Lu})(\p{Lu}\p{Ll})/gu;
exports.regExWords = /\p{L}(?:(?:\\?['’])?\p{L})*/gu;
exports.regExWordsAndDigits = /(?:\d+)?[\p{L}_'’-](?:(?:\\?['’])?[\p{L}\w'’-])*/gu;
exports.regExIgnoreCharacters = /\p{sc=Hiragana}|\p{sc=Han}|\p{sc=Katakana}|[\u30A0-\u30FF]|[\p{sc=Hangul}]/gu;
exports.regExFirstUpper = /^\p{Lu}\p{Ll}+$/u;
exports.regExAllUpper = /^\p{Lu}+$/u;
exports.regExAllLower = /^\p{Ll}+$/u;
exports.regExPossibleWordBreaks = /[_-]/g;
exports.regExMatchRegExParts = /^\/(.*)\/([gimuy]*)$/;
exports.regExAccents = /\p{M}/gu;
exports.regExEscapeCharacters = /(?<=\\)[anrvtbf]/gi;
exports.regExDanglingQuote = /(?<=\P{L}\p{L}?)[']/gu;
/** Match tailing endings after CAPS words */
exports.regExTrailingEndings = /(?<=\p{Lu}{2})['’]?(?:s|d|ing[s]|ies|e[ds]|ning|th|nth)(?!\p{Ll})/gu;
