// cspell-trie-lib/dist/esm/lib/mappers/mapDictionaryInfoToWeightMap.js
import { addAdjustment, createWeightMap } from '../distance/weightedMaps.js';
import { mapDictionaryInformation, mapDictionaryInformationToAdjustment } from './mapDictionaryInfo.js';
const defaultDefs = [
    {
        map: '1234567890-.',
        insDel: 1,
        penalty: 200,
    },
];
const defaultAdjustments = [
    {
        id: 'compound-case-change',
        regexp: /\p{Ll}∙\p{Lu}/gu,
        penalty: 1000,
    },
    {
        id: 'short-compounds-1',
        regexp: /^[^∙]{0,2}(?=∙)|∙[^∙]{0,2}(?=∙|$)/gm,
        penalty: 100,
    },
    {
        id: 'short-compounds-3',
        regexp: /^[^∙]{3}(?=∙)|∙[^∙]{3}(?=∙|$)/gm,
        penalty: 50,
    },
];
export function mapDictionaryInformationToWeightMap(dictInfo) {
    const defs = mapDictionaryInformation(dictInfo).concat(defaultDefs);
    const adjustments = mapDictionaryInformationToAdjustment(dictInfo);
    const map = createWeightMap(...defs);
    addAdjustment(map, ...defaultAdjustments, ...adjustments);
    return map;
}
export const __testing__ = {};
