// cspell-trie-lib/dist/esm/lib/index.mjs
export { consolidate } from './consolidate.mjs';
export { createWeightedMap, editDistance, editDistanceWeighted } from './distance/index.mjs';
export { importTrie, serializeTrie } from './io/importExport.mjs';
export { mapDictionaryInformationToWeightMap } from './mappers/mapDictionaryInfoToWeightMap.mjs';
export { createDictionaryLineParserMapper as createDictionaryLineParser, parseDictionary, parseDictionaryLines, } from './SimpleDictionaryParser.mjs';
export { impersonateCollector, suggestionCollector } from './suggestCollector.mjs';
export { CASE_INSENSITIVE_PREFIX, COMPOUND, COMPOUND_FIX, defaultTrieOptions, FORBID, FORBID_PREFIX, NORMALIZED, OPTIONAL_COMPOUND, OPTIONAL_COMPOUND_FIX, Trie, } from './trie.mjs';
export { countNodes, countWords, createTrieRoot, createTriFromList, findNode, has, insert, isCircular, isWordTerminationNode, iterateTrie, iteratorTrieWords, orderTrie, trieNodeToRoot, walk, } from './trie-util.mjs';
export { buildTrie, buildTrieFast, TrieBuilder } from './TrieBuilder.mjs';
export { ChildMap, FLAG_WORD } from './TrieNode.mjs';
export { isDefined } from './utils/isDefined.mjs';
export { mergeDefaults } from './utils/mergeDefaults.mjs';
export { mergeOptionalWithDefaults } from './utils/mergeOptionalWithDefaults.mjs';
export { normalizeWord, normalizeWordForCaseInsensitive, normalizeWordToLowercase } from './utils/normalizeWord.mjs';
export { expandCharacterSet } from './utils/text.mjs';
export { CompoundWordsMethod, hintedWalker, JOIN_SEPARATOR, walker, WORD_SEPARATOR } from './walker/index.mjs';
