// @cspell/cspell-service-bus/dist/bus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createServiceBus = exports.ServiceBus = void 0;
const errors_1 = require("./errors");
const request_1 = require("./request");
const MAX_DEPTH = 10;
class ServiceBus {
    constructor(handlers = []) {
        this.handlers = [];
        handlers.forEach((h) => this.addHandler(h));
    }
    addHandler(handler, name = 'anonymous', description) {
        const h = typeof handler === 'function' ? { fn: handler, name, description } : handler;
        const { fn, name: _name, description: _description } = h;
        this.handlers.push({ fn, name: _name, description: _description });
        return this;
    }
    dispatch(request) {
        let depth = 0;
        const dispatcher = { dispatch };
        const handler = this.reduceHandlers(this.handlers, request, dispatcher, this.defaultHandler);
        function dispatch(request) {
            ++depth;
            if (depth >= MAX_DEPTH) {
                return (0, request_1.createResponseFail)(request, new errors_1.ErrorServiceRequestDepthExceeded(request, depth));
            }
            const response = handler(request);
            --depth;
            return response;
        }
        return dispatch(request);
    }
    defaultHandler(request) {
        return (0, request_1.createResponseFail)(request, new errors_1.ErrorUnhandledRequest(request));
    }
    reduceHandlers(handlers, request, dispatcher, defaultHandler) {
        const _handlers = handlers.map((m) => ({ ...m, fn: m.fn(dispatcher) }));
        const handler = _handlers.reduce((next, h) => {
            const fn = h.fn(next);
            return (req) => {
                try {
                    return fn(req);
                }
                catch (e) {
                    return (0, request_1.createResponseFail)(request, new errors_1.UnhandledHandlerError(h.name, h.description, e));
                }
            };
        }, defaultHandler);
        return handler;
    }
}
exports.ServiceBus = ServiceBus;
function createServiceBus(handlers = []) {
    return new ServiceBus(handlers);
}
exports.createServiceBus = createServiceBus;
