// @cspell/cspell-service-bus/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorDuplicateSubsystem = exports.UnhandledHandlerError = exports.ErrorServiceRequestDepthExceeded = exports.ErrorUnhandledRequest = void 0;
class ErrorUnhandledRequest extends Error {
    constructor(request) {
        super(`Unhandled Request: ${request.type}`);
        this.request = request;
    }
}
exports.ErrorUnhandledRequest = ErrorUnhandledRequest;
class ErrorServiceRequestDepthExceeded extends Error {
    constructor(request, depth) {
        super(`Service Request Depth ${depth} Exceeded: ${request.type}`);
        this.request = request;
        this.depth = depth;
    }
}
exports.ErrorServiceRequestDepthExceeded = ErrorServiceRequestDepthExceeded;
class UnhandledHandlerError extends Error {
    constructor(handlerName, handlerDescription, cause) {
        super(`Unhandled Error in Handler: ${handlerName}`);
        this.handlerName = handlerName;
        this.handlerDescription = handlerDescription;
        this.cause = cause;
    }
}
exports.UnhandledHandlerError = UnhandledHandlerError;
class ErrorDuplicateSubsystem extends Error {
    constructor(subsystemName) {
        super(`Duplicate subsystem: ${subsystemName}`);
        this.subsystemName = subsystemName;
    }
}
exports.ErrorDuplicateSubsystem = ErrorDuplicateSubsystem;
