// @cspell/cspell-service-bus/dist/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.__testing__ = exports.isInstanceOfFn = exports.isServiceResponseFailure = exports.isServiceResponseSuccess = exports.createResponseFail = exports.createResponse = exports.ServiceRequest = void 0;
class BaseServiceRequest {
    constructor(type, params) {
        this.type = type;
        this.params = params;
    }
}
class ServiceRequest extends BaseServiceRequest {
    constructor(type, params) {
        super(type, params);
    }
}
exports.ServiceRequest = ServiceRequest;
function createResponse(value, _req) {
    return { value };
}
exports.createResponse = createResponse;
function createResponseFail(_request, error) {
    return { error };
}
exports.createResponseFail = createResponseFail;
function isServiceResponseSuccess(res) {
    return 'value' in res && res.error === undefined;
}
exports.isServiceResponseSuccess = isServiceResponseSuccess;
function isServiceResponseFailure(res) {
    return res.error !== undefined;
}
exports.isServiceResponseFailure = isServiceResponseFailure;
function isInstanceOfFn(constructor) {
    return (t) => t instanceof constructor;
}
exports.isInstanceOfFn = isInstanceOfFn;
exports.__testing__ = {
    BaseServiceRequest,
};
