// @cspell/cspell-service-bus/dist/SystemServiceBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createSubsystemServiceBus = exports.RequestCreateSubsystemFactory = exports.RequestRegisterHandlerFactory = exports.createSystemServiceBus = void 0;
const assert_1 = require("./assert");
const bus_1 = require("./bus");
const createRequestHandler_1 = require("./createRequestHandler");
const request_1 = require("./request");
const requestFactory_1 = require("./requestFactory");
const errors_1 = require("./errors");
class SystemServiceBusImpl {
    constructor() {
        this.serviceBus = (0, bus_1.createServiceBus)();
        this._subsystems = new Map();
        this.bindDefaultHandlers();
        this.createSubsystem('Default Subsystem', '' /* match everything */);
    }
    bindDefaultHandlers() {
        this.serviceBus.addHandler((0, createRequestHandler_1.createRequestHandler)(exports.RequestCreateSubsystemFactory, (req) => {
            const { name, requestPattern } = req.params;
            if (this._subsystems.has(name)) {
                throw new errors_1.ErrorDuplicateSubsystem(name);
            }
            const sub = createSubsystemServiceBus(name, requestPattern);
            this._subsystems.set(name, sub);
            this.serviceBus.addHandler(sub.handler);
            return (0, request_1.createResponse)(sub);
        }));
    }
    dispatch(request) {
        return this.serviceBus.dispatch(request);
    }
    createSubsystem(name, requestPattern) {
        const res = this.dispatch(exports.RequestCreateSubsystemFactory.create({ name, requestPattern }));
        (0, assert_1.assert)(res?.value);
        return res.value;
    }
    registerHandler(requestPrefix, handler) {
        const request = exports.RequestRegisterHandlerFactory.create({ requestPrefix, handler });
        this.serviceBus.dispatch(request);
        return this;
    }
    registerRequestHandler(requestDef, fn, name, description) {
        this.registerHandler(requestDef.type, (0, createRequestHandler_1.createRequestHandler)(requestDef, fn, name, description));
        return this;
    }
    get subsystems() {
        return new Map(this._subsystems);
    }
}
function createSystemServiceBus() {
    return new SystemServiceBusImpl();
}
exports.createSystemServiceBus = createSystemServiceBus;
const TypeRequestRegisterHandler = 'System:RegisterHandler';
exports.RequestRegisterHandlerFactory = (0, requestFactory_1.requestFactory)(TypeRequestRegisterHandler);
const TypeRequestCreateSubsystem = 'System:CreateSubsystem';
exports.RequestCreateSubsystemFactory = (0, requestFactory_1.requestFactory)(TypeRequestCreateSubsystem);
class SubsystemServiceBusImpl extends bus_1.ServiceBus {
    constructor(name, requestPattern) {
        super();
        this.name = name;
        this.requestPattern = requestPattern;
        this.canHandleType =
            typeof requestPattern === 'string'
                ? (reqType) => reqType.startsWith(requestPattern)
                : (reqType) => requestPattern.test(reqType);
        const handleRegistration = (0, createRequestHandler_1.createRequestHandler)(exports.RequestRegisterHandlerFactory, (req, next) => this.handleRegistrationReq(req, next), 'Subsystem Register Handlers for ' + name, `Matches against: <${requestPattern.toString()}>`);
        this.addHandler(handleRegistration);
        this.handler = {
            name: 'Subsystem: ' + name,
            description: `Process Requests Matching: <${requestPattern.toString()}>`,
            fn: (dispatcher) => this._handler(dispatcher),
        };
    }
    handleRegistrationReq(request, next) {
        // console.log(`${this.name}.handleRegistrationReq %o`, request);
        if (!this.canHandleType(request.params.requestPrefix)) {
            // console.log(`${this.name}.handleRegistrationReq skip`);
            return next(request);
        }
        // console.log(`${this.name}.handleRegistrationReq add ***`);
        this.addHandler(request.params.handler);
        return (0, request_1.createResponse)(this);
    }
    _handler(dispatcher) {
        return (next) => (req) => {
            if (!this.canHandleType(req.type) && !exports.RequestRegisterHandlerFactory.is(req))
                return next(req);
            const dispatch = this.reduceHandlers(this.handlers, req, dispatcher, next);
            return dispatch(req);
        };
    }
}
function createSubsystemServiceBus(name, requestPattern) {
    return new SubsystemServiceBusImpl(name, requestPattern);
}
exports.createSubsystemServiceBus = createSubsystemServiceBus;
