// @cspell/cspell-pipe/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.helpers = exports.operators = exports.pipeSync = exports.pipeAsync = exports.opUnique = exports.opMap = exports.opJoinStrings = exports.opFlatten = exports.opFilter = exports.opAwaitAsync = exports.toAsyncIterable = exports.toArray = exports.isAsyncIterable = void 0;
const _helpers = __importStar(require("./helpers"));
const _operators = __importStar(require("./operators"));
var helpers_1 = require("./helpers");
Object.defineProperty(exports, "isAsyncIterable", { enumerable: true, get: function () { return helpers_1.isAsyncIterable; } });
Object.defineProperty(exports, "toArray", { enumerable: true, get: function () { return helpers_1.toArray; } });
Object.defineProperty(exports, "toAsyncIterable", { enumerable: true, get: function () { return helpers_1.toAsyncIterable; } });
var operators_1 = require("./operators");
Object.defineProperty(exports, "opAwaitAsync", { enumerable: true, get: function () { return operators_1.opAwaitAsync; } });
Object.defineProperty(exports, "opFilter", { enumerable: true, get: function () { return operators_1.opFilter; } });
Object.defineProperty(exports, "opFlatten", { enumerable: true, get: function () { return operators_1.opFlatten; } });
Object.defineProperty(exports, "opJoinStrings", { enumerable: true, get: function () { return operators_1.opJoinStrings; } });
Object.defineProperty(exports, "opMap", { enumerable: true, get: function () { return operators_1.opMap; } });
Object.defineProperty(exports, "opUnique", { enumerable: true, get: function () { return operators_1.opUnique; } });
var pipe_1 = require("./pipe");
Object.defineProperty(exports, "pipeAsync", { enumerable: true, get: function () { return pipe_1.pipeAsync; } });
Object.defineProperty(exports, "pipeSync", { enumerable: true, get: function () { return pipe_1.pipeSync; } });
exports.operators = _operators;
exports.helpers = _helpers;
