// cspell/dist/esm/link.mjs
import chalk from 'chalk';
import { Link } from 'cspell-lib';
export const listGlobalImports = Link.listGlobalImports;
export const addPathsToGlobalImports = Link.addPathsToGlobalImports;
export const removePathsFromGlobalImports = Link.removePathsFromGlobalImports;
export function listGlobalImportsResultToTable(results) {
    const header = ['id', 'package', 'name', 'filename', 'dictionaries', 'errors'];
    const decorate = (isError) => (isError ? (s) => chalk.red(s) : (s) => s);
    function toColumns(r) {
        return [
            r.id,
            r.package?.name,
            r.name,
            r.filename,
            r.dictionaryDefinitions?.map((def) => def.name).join(', '),
            r.error ? 'Failed to read file.' : '',
        ]
            .map((c) => c || '')
            .map(decorate(!!r.error));
    }
    return {
        header,
        rows: results.map(toColumns),
    };
}
export function addPathsToGlobalImportsResultToTable(results) {
    const header = ['filename', 'errors'];
    const decorate = (isError) => (isError ? (s) => chalk.red(s) : (s) => s);
    function toColumns(r) {
        return [r.resolvedToFilename || r.filename, r.error ? 'Failed to read file.' : '']
            .map((c) => c || '')
            .map(decorate(!!r.error));
    }
    return {
        header,
        rows: results.resolvedSettings.map(toColumns),
    };
}
