// cspell-dictionary/dist/esm/SpellingDictionary/SpellingDictionaryFromTrie.mjs
import { opConcatMap, pipe } from '@cspell/cspell-pipe/sync';
import { CompoundWordsMethod, importTrie, suggestionCollector, Trie } from 'cspell-trie-lib';
import { autoCache, createCache01 } from '../util/AutoCache.mjs';
import { clean } from '../util/clean.mjs';
import { createMapper, createRepMapper } from '../util/repMap.mjs';
import * as Defaults from './defaults.mjs';
import { createWeightMapFromDictionaryInformation, defaultNumSuggestions, hasOptionToSearchOption, impersonateCollector, suggestArgsToSuggestOptions, wordSearchForms, wordSuggestFormsArray, } from './SpellingDictionaryMethods.mjs';
const findWordOptionsCaseSensitive = Object.freeze({ caseSensitive: true });
const findWordOptionsNotCaseSensitive = Object.freeze({ caseSensitive: false });
export class SpellingDictionaryFromTrie {
    constructor(trie, name, options, source = 'from trie', size) {
        this.trie = trie;
        this.name = name;
        this.options = options;
        this.source = source;
        this._size = 0;
        this.knownWords = new Set();
        this.unknownWords = new Set();
        this.type = 'SpellingDictionaryFromTrie';
        this._find = findCache((word, useCompounds, ignoreCase) => this.findAnyForm(word, useCompounds, ignoreCase));
        this._isForbidden = autoCache((word) => {
            return this.trie.isForbiddenWord(word);
        });
        this.mapWord = createMapper(options.repMap, options.dictionaryInformation?.ignore);
        this.remapWord = createRepMapper(options.repMap, options.dictionaryInformation?.ignore);
        this.isDictionaryCaseSensitive = options.caseSensitive ?? !trie.isLegacy;
        this.containsNoSuggestWords = options.noSuggest || false;
        this._size = size || 0;
        this.weightMap = options.weightMap || createWeightMapFromDictionaryInformation(options.dictionaryInformation);
    }
    get size() {
        if (!this._size) {
            // walk the trie and get the approximate size.
            const i = this.trie.iterate();
            let deeper = true;
            let size = 0;
            for (let r = i.next(); !r.done; r = i.next(deeper)) {
                // count all nodes even though they are not words.
                // because we are not going to all the leaves, this should give a good enough approximation.
                size += 1;
                deeper = r.value.text.length < 5;
            }
            this._size = size;
        }
        return this._size;
    }
    has(word, hasOptions) {
        const { useCompounds, ignoreCase } = this.resolveOptions(hasOptions);
        const r = this._find(word, useCompounds, ignoreCase);
        return !!r && !r.forbidden && !!r.found;
    }
    find(word, hasOptions) {
        const { useCompounds, ignoreCase } = this.resolveOptions(hasOptions);
        const r = this._find(word, useCompounds, ignoreCase);
        const { forbidden = this.isForbidden(word) } = r || {};
        if (!r && !forbidden)
            return undefined;
        const { found = forbidden ? word : false } = r || {};
        const noSuggest = found !== false && this.containsNoSuggestWords;
        return { found, forbidden, noSuggest };
    }
    resolveOptions(hasOptions) {
        const { useCompounds = this.options.useCompounds, ignoreCase = Defaults.ignoreCase } = hasOptionToSearchOption(hasOptions);
        return { useCompounds, ignoreCase };
    }
    findAnyForm(word, useCompounds, ignoreCase) {
        const outerForms = outerWordForms(word, this.remapWord ? this.remapWord : (word) => [this.mapWord(word)]);
        for (const form of outerForms) {
            const r = this._findAnyForm(form, useCompounds, ignoreCase);
            if (r)
                return r;
        }
        return undefined;
    }
    _findAnyForm(mWord, useCompounds, ignoreCase) {
        const opts = ignoreCase ? findWordOptionsNotCaseSensitive : findWordOptionsCaseSensitive;
        const findResult = this.trie.findWord(mWord, opts);
        if (findResult.found !== false) {
            return findResult;
        }
        const forms = wordSearchForms(mWord, this.isDictionaryCaseSensitive, ignoreCase);
        for (const w of forms) {
            const findResult = this.trie.findWord(w, opts);
            if (findResult.found !== false) {
                return findResult;
            }
        }
        if (useCompounds) {
            const optsUseCompounds = { ...opts, useLegacyWordCompounds: useCompounds };
            for (const w of forms) {
                const findResult = this.trie.findWord(w, optsUseCompounds);
                if (findResult.found !== false) {
                    return findResult;
                }
            }
        }
        return undefined;
    }
    isNoSuggestWord(word, options) {
        return this.containsNoSuggestWords ? this.has(word, options) : false;
    }
    isForbidden(word, _ignoreCaseAndAccents) {
        return this._isForbidden(word);
    }
    suggest(...args) {
        const [word] = args;
        const suggestOptions = suggestArgsToSuggestOptions(args);
        return this._suggest(word, suggestOptions);
    }
    _suggest(word, suggestOptions) {
        const { numSuggestions = defaultNumSuggestions, numChanges, includeTies, ignoreCase, timeout } = suggestOptions;
        function filter(_word) {
            return true;
        }
        const collector = suggestionCollector(word, clean({
            numSuggestions,
            filter,
            changeLimit: numChanges,
            includeTies,
            ignoreCase,
            timeout,
            weightMap: this.weightMap,
        }));
        this.genSuggestions(collector, suggestOptions);
        return collector.suggestions.map((r) => ({ ...r, word: r.word }));
    }
    genSuggestions(collector, suggestOptions) {
        if (this.options.noSuggest)
            return;
        const _compoundMethod = suggestOptions.compoundMethod ??
            (this.options.useCompounds ? CompoundWordsMethod.JOIN_WORDS : CompoundWordsMethod.NONE);
        wordSuggestFormsArray(collector.word).forEach((w) => this.trie.genSuggestions(impersonateCollector(collector, w), _compoundMethod));
    }
    getErrors() {
        return [];
    }
}
SpellingDictionaryFromTrie.cachedWordsLimit = 50000;
/**
 * Create a dictionary from a trie file.
 * @param data - contents of a trie file.
 * @param name - name of dictionary
 * @param source - filename or uri
 * @param options - options.
 * @returns SpellingDictionary
 */
export function createSpellingDictionaryFromTrieFile(data, name, source, options) {
    data = typeof data === 'string' ? data.split('\n') : data;
    const trieNode = importTrie(data);
    const trie = new Trie(trieNode);
    return new SpellingDictionaryFromTrie(trie, name, options, source);
}
function findCache(fn, size = 2000) {
    const cache = createCache01(size);
    function find(word, useCompounds, ignoreCase) {
        const r = cache.get(word);
        if (r !== undefined) {
            if (r.useCompounds === useCompounds && r.ignoreCase === ignoreCase) {
                return r.findResult;
            }
        }
        const findResult = fn(word, useCompounds, ignoreCase);
        cache.set(word, { useCompounds, ignoreCase, findResult });
        return findResult;
    }
    return find;
}
function outerWordForms(word, mapWord) {
    const forms = pipe([word], opConcatMap((word) => [word, word.normalize('NFC'), word.normalize('NFD')]), opConcatMap((word) => [word, ...mapWord(word)]));
    return new Set(forms);
}
export const __testing__ = { outerWordForms };
