// cspell-lib/dist/esm/Settings/Controller/pnpLoader.mjs
/**
 * Handles loading of `.pnp.js` and `.pnp.js` files.
 */
import clearModule from 'clear-module';
import findUp from 'find-up';
import importFresh from 'import-fresh';
import { toUri, uriToFilePath } from '../../util/Uri.mjs';
import { UnsupportedPnpFile } from './ImportError.mjs';
const defaultPnpFiles = ['.pnp.cjs', '.pnp.js'];
const supportedSchemas = new Set(['file']);
const cachedRequests = new Map();
let lock = undefined;
const cachedPnpImportsSync = new Map();
const cachedRequestsSync = new Map();
export class PnpLoader {
    constructor(pnpFiles = defaultPnpFiles) {
        this.pnpFiles = pnpFiles;
        this.cacheKeySuffix = ':' + pnpFiles.join();
    }
    /**
     * Request that the nearest .pnp file gets loaded
     * @param uriDirectory starting directory
     * @returns promise - rejects on error - success if loaded or not found.
     */
    async load(uriDirectory) {
        if (!supportedSchemas.has(uriDirectory.scheme))
            return undefined;
        await lock;
        const cacheKey = this.calcKey(uriDirectory);
        const cached = cachedRequests.get(cacheKey);
        if (cached)
            return cached;
        const r = findPnpAndLoad(uriDirectory, this.pnpFiles);
        cachedRequests.set(cacheKey, r);
        const result = await r;
        cachedRequestsSync.set(cacheKey, result);
        return result;
    }
    async peek(uriDirectory) {
        if (!supportedSchemas.has(uriDirectory.scheme))
            return undefined;
        await lock;
        const cacheKey = this.calcKey(uriDirectory);
        return cachedRequests.get(cacheKey) ?? Promise.resolve(undefined);
    }
    /**
     * Request that the nearest .pnp file gets loaded
     * @param uriDirectory starting directory
     * @returns promise - rejects on error - success if loaded or not found.
     */
    loadSync(uriDirectory) {
        if (!supportedSchemas.has(uriDirectory.scheme))
            return undefined;
        const cacheKey = this.calcKey(uriDirectory);
        const cached = cachedRequestsSync.get(cacheKey);
        if (cached)
            return cached;
        const r = findPnpAndLoadSync(uriDirectory, this.pnpFiles);
        cachedRequestsSync.set(cacheKey, r);
        cachedRequests.set(cacheKey, Promise.resolve(r));
        return r;
    }
    peekSync(uriDirectory) {
        if (!supportedSchemas.has(uriDirectory.scheme))
            return undefined;
        const cacheKey = this.calcKey(uriDirectory);
        return cachedRequestsSync.get(cacheKey);
    }
    /**
     * Clears the cached so .pnp files will get reloaded on request.
     */
    clearCache() {
        return clearPnPGlobalCache();
    }
    calcKey(uriDirectory) {
        return uriDirectory.toString() + this.cacheKeySuffix;
    }
}
export function pnpLoader(pnpFiles) {
    return new PnpLoader(pnpFiles);
}
/**
 * @param uriDirectory - directory to start at.
 */
async function findPnpAndLoad(uriDirectory, pnpFiles) {
    const found = await findUp(pnpFiles, { cwd: uriToFilePath(uriDirectory) });
    return loadPnpIfNeeded(found);
}
/**
 * @param uriDirectory - directory to start at.
 */
function findPnpAndLoadSync(uriDirectory, pnpFiles) {
    const found = findUp.sync(pnpFiles, { cwd: uriToFilePath(uriDirectory) });
    return loadPnpIfNeeded(found);
}
function loadPnpIfNeeded(found) {
    if (!found)
        return undefined;
    const c = cachedPnpImportsSync.get(found);
    if (c || cachedPnpImportsSync.has(found))
        return c;
    const r = loadPnp(found);
    cachedPnpImportsSync.set(found, r);
    return r;
}
function loadPnp(pnpFile) {
    const pnp = importFresh(pnpFile);
    if (pnp.setup) {
        pnp.setup();
        return toUri(pnpFile);
    }
    throw new UnsupportedPnpFile(`Unsupported pnp file: "${pnpFile}"`);
}
export function clearPnPGlobalCache() {
    if (lock)
        return lock;
    lock = _cleanCache().finally(() => {
        lock = undefined;
    });
    return lock;
}
async function _cleanCache() {
    await Promise.all([...cachedRequests.values()].map(rejectToUndefined));
    const modules = [...cachedPnpImportsSync.values()];
    modules.forEach((r) => r && clearModule.single(uriToFilePath(r)));
    cachedRequests.clear();
    cachedRequestsSync.clear();
    cachedPnpImportsSync.clear();
    return undefined;
}
function rejectToUndefined(p) {
    return p.catch(() => undefined);
}
