// cspell-lib/dist/cjs/FeatureFlags/FeatureFlags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSystemFeatureFlags = exports.UnknownFeatureFlagError = exports.FeatureFlags = void 0;
let systemFeatureFlags;
/**
 * Feature Flags are used to turn on/off features.
 * These are primarily used before a feature has been fully released.
 */
class FeatureFlags {
    constructor(flags = []) {
        this.flagValues = new Map();
        this.flags = new Map(flags.map((f) => [f.name, f]));
    }
    register(flagOrName, description) {
        if (typeof flagOrName === 'string') {
            return this.register({ name: flagOrName, description: description || '' });
        }
        this.flags.set(flagOrName.name, flagOrName);
        return this;
    }
    getFlag(flag) {
        return this.flagValues.get(flag);
    }
    getFlagBool(flag) {
        return toBool(this.getFlag(flag));
    }
    setFlag(flag, value = true) {
        if (!this.flags.has(flag)) {
            throw new UnknownFeatureFlagError(flag);
        }
        this.flagValues.set(flag, value);
        return this;
    }
    getFlagInfo(flag) {
        return this.flags.get(flag);
    }
    getFlags() {
        return [...this.flags.values()];
    }
    getFlagValues() {
        return new Map(this.flagValues);
    }
    reset() {
        this.flagValues.clear();
        return this;
    }
}
exports.FeatureFlags = FeatureFlags;
class UnknownFeatureFlagError extends Error {
    constructor(flag) {
        super(`Unknown feature flag: ${flag}`);
        this.flag = flag;
    }
}
exports.UnknownFeatureFlagError = UnknownFeatureFlagError;
function getSystemFeatureFlags() {
    return systemFeatureFlags || (systemFeatureFlags = new FeatureFlags());
}
exports.getSystemFeatureFlags = getSystemFeatureFlags;
const boolValues = {
    0: false,
    1: true,
    f: false,
    false: false,
    n: false,
    no: false,
    t: true,
    true: true,
    y: true,
    yes: true,
};
function toBool(value) {
    if (typeof value !== 'string')
        return value;
    return boolValues[value.toLowerCase()];
}
